<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>OrdinalGWAS.jl · OrdinalGWAS</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>OrdinalGWAS</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>OrdinalGWAS.jl</a><ul class="internal"><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Example-data-sets"><span>Example data sets</span></a></li><li><a class="tocitem" href="#Basic-usage"><span>Basic usage</span></a></li><li><a class="tocitem" href="#VCF-Formatted-Files"><span>VCF Formatted Files</span></a></li><li><a class="tocitem" href="#BGEN-Formatted-Files"><span>BGEN Formatted Files</span></a></li><li><a class="tocitem" href="#Link-functions"><span>Link functions</span></a></li><li><a class="tocitem" href="#SNP-models"><span>SNP models</span></a></li><li><a class="tocitem" href="#SNP-and/or-sample-masks"><span>SNP and/or sample masks</span></a></li><li><a class="tocitem" href="#Likelihood-ratio-test-(LRT)"><span>Likelihood ratio test (LRT)</span></a></li><li><a class="tocitem" href="#Score-test-for-screening,-LRT-for-power"><span>Score test for screening, LRT for power</span></a></li><li><a class="tocitem" href="#GxE-or-other-interactions"><span>GxE or other interactions</span></a></li><li><a class="tocitem" href="#SNP-set-testing"><span>SNP-set testing</span></a></li><li><a class="tocitem" href="#Plotting-Results"><span>Plotting Results</span></a></li><li><a class="tocitem" href="#Docker"><span>Docker</span></a></li><li><a class="tocitem" href="#Multiple-Plink-file-sets"><span>Multiple Plink file sets</span></a></li><li><a class="tocitem" href="#Multiple-Plink-file-sets-on-cluster"><span>Multiple Plink file sets on cluster</span></a></li><li><a class="tocitem" href="#Proportional-Odds-Assumption"><span>Proportional Odds Assumption</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>OrdinalGWAS.jl</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>OrdinalGWAS.jl</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/OrdinalGWAS.jl/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="OrdinalGWAS.jl"><a class="docs-heading-anchor" href="#OrdinalGWAS.jl">OrdinalGWAS.jl</a><a id="OrdinalGWAS.jl-1"></a><a class="docs-heading-anchor-permalink" href="#OrdinalGWAS.jl" title="Permalink"></a></h1><p>OrdinalGWAS.jl is a Julia package for performing genome-wide association studies (GWAS) for ordered categorical phenotypes using <a href="https://en.wikipedia.org/wiki/Ordered_logit">proportional odds model</a> or <a href="https://en.wikipedia.org/wiki/Ordered_probit">ordred Probit model</a>. It is useful when the phenotype takes ordered discrete values, e.g., disease status (undiagnosed, pre-disease, mild, moderate, severe). </p><p>The methods and applications of this software package are detailed in the following publication:</p><p><em>German CA, Sinsheimer JS, Klimentidis YC, Zhou H, Zhou JJ. (2020) Ordered multinomial regression for genetic association analysis of ordinal phenotypes at Biobank scale. Genetic Epidemiology. 44:248-260. <a href="https://doi.org/10.1002/gepi.22276">https://doi.org/10.1002/gepi.22276</a></em></p><p>OrdinalGWAS.jl currently supports <a href="https://zzz.bwh.harvard.edu/plink/">PLINK</a>, <a href="https://en.wikipedia.org/wiki/Variant_Call_Format">VCF</a> (both dosage and genotype data) file formats, and <a href="https://www.well.ox.ac.uk/~gav/bgen_format/">BGEN</a> file formats. We plan to add <a href="https://www.cog-genomics.org/plink/2.0/formats#pgen">PGEN</a> support in the future. </p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>This package requires Julia v1.6 or later. The package has been registered and can be installed using Julia package manager. Start julia and type:</p><pre><code class="language-julia hljs">using Pkg
Pkg.add(&quot;OrdinalGWAS&quot;)</code></pre><pre><code class="language-julia hljs"># machine information for this tutorial
versioninfo()</code></pre><pre><code class="nohighlight hljs">Julia Version 1.6.2
Commit 1b93d53fc4 (2021-07-14 15:36 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)</code></pre><pre><code class="language-julia hljs"># for use in this tutorial
using BenchmarkTools, CSV, Glob, SnpArrays, OrdinalGWAS, DataFrames</code></pre><h2 id="Example-data-sets"><a class="docs-heading-anchor" href="#Example-data-sets">Example data sets</a><a id="Example-data-sets-1"></a><a class="docs-heading-anchor-permalink" href="#Example-data-sets" title="Permalink"></a></h2><p>The <code>data</code> folder of the package contains the example data sets for use with PLINK and VCF Files. In general, the user can locate this folder by command:</p><pre><code class="language-julia hljs">using OrdinalGWAS
const datadir = normpath(joinpath(dirname(pathof(OrdinalGWAS)), &quot;../data/&quot;))</code></pre><pre><code class="nohighlight hljs">&quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/&quot;</code></pre><pre><code class="language-julia hljs"># content of the data folder
readdir(glob&quot;*.*&quot;, datadir)</code></pre><pre><code class="nohighlight hljs">16-element Vector{String}:
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/bgen_ex.csv&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/bgen_snpsetfile.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/bgen_test.bgen&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/bgen_test.bgen.bgi&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/covariate.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.map&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap_snpsetfile.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/simtrait.jl&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/simtrait_bgen.jl&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/simtrait_vcf.jl&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/snpsetfile_vcf.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/vcf_example.csv&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/vcf_test.vcf.gz&quot;</code></pre><p>The <code>hapmap3</code> files <code>covariate.txt</code> file correspond to data examples using PLINK formatted files (.bed, .bim, .fam). </p><p>The <code>vcf_test.vcf.gz</code> and <code>vcf_example.csv</code> files are for an example analysis using VCF formatted files. </p><h2 id="Basic-usage"><a class="docs-heading-anchor" href="#Basic-usage">Basic usage</a><a id="Basic-usage-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-usage" title="Permalink"></a></h2><p>The following command performs GWAS using the <a href="https://en.wikipedia.org/wiki/Ordered_logit">proportional odds model</a> for the hapmap3 PLINK files. The output is the fitted null model.</p><p>The default type of GWAS performed is a single-snp significance genome-wide scan, this can be changed by the keyword <code>analysistype</code> (default is &quot;singlesnp&quot;). Other types of analyses are gone over later. </p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><p>For documentation of the <code>ordinalgwas</code> function, type <code>?ordinalgwas</code> in Julia REPL.</p><article class="docstring"><header><a class="docstring-binding" id="OrdinalGWAS.ordinalgwas" href="#OrdinalGWAS.ordinalgwas"><code>OrdinalGWAS.ordinalgwas</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">ordinalgwas(nullformula, covfile, geneticfile; kwargs...)
ordinalgwas(nullformula, df, geneticfile; kwargs...)
ordinalgwas(fittednullmodel, geneticfile; kwargs...)
ordinalgwas(fittednullmodel, bedfile, bimfile, bedn; kwargs...)</code></pre><p><strong>Positional arguments</strong></p><ul><li><code>nullformula::FormulaTerm</code>: formula for the null model.</li><li><code>covfile::AbstractString</code>: covariate file (csv) with one header line. One column    should be the ordinal phenotype coded as integers starting from 1.  For example,    ordinal phenotypes can be coded as 1, 2, 3, 4 but not 0, 1, 2, 3.  </li><li><code>df::DataFrame</code>: DataFrame containing response and regressors for null model.</li><li><code>geneticfile::Union{Nothing, AbstractString}</code>: File containing genetic information for GWAS.   This includes a PLINK file name without the .bed, .fam, or .bim    extensions or a VCF file without the .vcf extension. If <code>geneticfile==nothing</code>,    only null model is fitted. If <code>geneticfile</code> is provided, bed, bim, and fam file (or vcf) with    the same <code>geneticfile</code> prefix need to exist. Compressed file formats such as gz and bz2    are allowed. Check all allowed formats by <code>SnpArrays.ALLOWED_FORMAT</code>. If you&#39;re using a VCF file,   make sure to use the <code>geneticformat = &quot;VCF&quot;</code> keyword option, and specificy dosage (:DS) or    genotype (:GT) data with the <code>vcftype</code> command.</li><li><code>fittednullmodel::StatsModels.TableRegressionModel</code>: the fitted null model    output from <code>ordinalgwas(nullformula, covfile)</code> or <code>ordinalgwas(nullformula, df)</code>.</li><li><code>bedfile::Union{AbstractString,IOStream}</code>: path to Plink bed file with full file name.</li><li><code>bimfile::Union{AbstractString,IOStream}</code>: path to Plink bim file with full file name.</li><li><code>bedn::Integer</code>: number of samples in bed/vcf file.</li></ul><p><strong>Keyword arguments</strong></p><ul><li><code>analysistype</code>::AbstractString: Type of analysis to conduct. Default is <code>singlesnp</code>. Other options are <code>snpset</code> and <code>gxe</code>.</li><li><code>geneticformat</code>::AbstractString: Type of file used for the genetic analysis. <code>&quot;PLINK&quot;</code>, <code>&quot;VCF&quot;</code>, and <code>&quot;BGEN&quot;</code> are currently supported. Default is PLINK.</li><li><code>vcftype</code>::Union{Symbol, Nothing}: Data to extract from the VCF file for the GWAS analysis. <code>:DS</code> for dosage or <code>:GT</code> for genotypes. Default is nothing.</li><li><code>nullfile::Union{AbstractString, IOStream}</code>: output file for the fitted null model;    default is <code>ordinalgwas.null.txt</code>. </li><li><code>pvalfile::Union{AbstractString, IOStream}</code>: output file for the gwas p-values; default is    <code>ordinalgwas.pval.txt</code>. </li><li><code>covtype::Vector{DataType}</code>: type information for <code>covfile</code>. This is useful   when <code>CSV.read(covarfile)</code> has parsing errors.  </li><li><code>covrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for covariate file.  </li><li><code>testformula::FormulaTerm</code>: formula for test unit. Default is <code>@formula(trait ~ 0 + snp)</code>.</li><li><code>test::Symbol</code>: <code>:score</code> (default) or <code>:lrt</code>.  </li><li><code>link::GLM.Link</code>: <code>LogitLink()</code> (default), <code>ProbitLink()</code>, <code>CauchitLink()</code>,   or <code>CloglogLink()</code>.</li><li><code>snpmodel</code>: <code>ADDITIVE_MODEL</code> (default), <code>DOMINANT_MODEL</code>, or <code>RECESSIVE_MODEL</code>.</li><li><code>snpinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: SNP indices for bed/vcf file.</li><li><code>geneticrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for bed/vcf file.</li><li><code>solver</code>: an optimizer supported by MathOptInterface. Default is    <code>NLopt.Optimizer()</code> with <code>algorithm=:LD_SLSQP</code> and <code>maxeval=4000</code>. Another common choice is    <code>Ipopt.Optimizer()</code>.</li><li><code>verbose::Bool</code>: default is <code>false</code>.</li><li><code>snpset::Union{Nothing, Integer, AbstractString, AbstractVector{&lt;:Integer}}</code>: Only include    if you are conducting a snpset analysis. An integer indicates a window of SNPs    (i.e. every 500 snps). An abstract string allows you to specify an input file,    with no header and two columns separated by a space. The first column must contain the snpset ID   and the second column must contain the snpid&#39;s identical to the bimfile. An AbstractVector   allows you to specify the snps you want to perform one joint snpset test for.</li><li><code>e::Union{AbstractString,Symbol}</code>: Only include if you are conducting a GxE analysis.    Enviromental variable to be used to test the GxE interaction.   For instance, for testing <code>sex &amp; snp</code> interaction, use <code>:sex</code> or <code>&quot;sex&quot;</code>.</li></ul><p><strong>Examples</strong></p><p>The following is an example of basic GWAS with PLINK files:</p><pre><code class="language-julia hljs">plinkfile = &quot;plinkexample&quot;
covfile = &quot;covexample&quot;
ordinalgwas(@formula(trait ~ sex), covfile, plkfile)</code></pre><p>The following is an example of basic GWAS with a VCF file using dosages then genotypes:</p><pre><code class="language-julia hljs">vcffile = &quot;vcfexample&quot;
covfile = &quot;covexample&quot;
ordinalgwas(@formula(trait ~ sex), covfile, vcffile; 
    geneticfile = &quot;VCF&quot;, vcftype = :DS)

ordinalgwas(@formula(trait ~ sex), covfile, vcffile; 
    geneticfile = &quot;VCF&quot;, vcftype = :GT)</code></pre><p>The following is an example of snpset GWAS (every 50 snps). For more types of snpset analyses see documentation:</p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), covfile, plkfile; 
    analysistype = &quot;snpset&quot;, snpset = 50)</code></pre><p>The following is an example of GxE GWAS testing the interaction effect:</p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), covfile, plkfile;
    analysistype = &quot;gxe&quot;, e = :sex)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/OrdinalGWAS.jl/blob/e0392cc19fb1fe3801c58d0e656dc73f792c3ea0/src/gwas.jl#L9-L98">source</a></section></article><h3 id="Formula-for-null-model"><a class="docs-heading-anchor" href="#Formula-for-null-model">Formula for null model</a><a id="Formula-for-null-model-1"></a><a class="docs-heading-anchor-permalink" href="#Formula-for-null-model" title="Permalink"></a></h3><p>The first argument specifies the null model without SNP effects, e.g., <code>@formula(trait ~ sex)</code>.</p><h3 id="Input-files"><a class="docs-heading-anchor" href="#Input-files">Input files</a><a id="Input-files-1"></a><a class="docs-heading-anchor-permalink" href="#Input-files" title="Permalink"></a></h3><p><code>ordinalgwas</code> expects two input files: one for responses plus covariates (second argument), the other the genetic file(s) for dosages/genotypes (third argument).</p><h4 id="Covariate-and-trait-file"><a class="docs-heading-anchor" href="#Covariate-and-trait-file">Covariate and trait file</a><a id="Covariate-and-trait-file-1"></a><a class="docs-heading-anchor-permalink" href="#Covariate-and-trait-file" title="Permalink"></a></h4><p>Covariates and phenotype are provided in a csv file, e.g., <code>covariate.txt</code>, which has one header line for variable names. In this example, variable <code>trait</code> is the ordered categorical phenotypes coded as integers 1 to 4. We want to include variable <code>sex</code> as the covariate in GWAS.</p><pre><code class="language-julia hljs">run(`head $(datadir)covariate.txt`);</code></pre><pre><code class="nohighlight hljs">famid,perid,faid,moid,sex,trait
2431,NA19916,0,0,1,4
2424,NA19835,0,0,2,4
2469,NA20282,0,0,2,4
2368,NA19703,0,0,1,3
2425,NA19901,0,0,2,3
2427,NA19908,0,0,1,4
2430,NA19914,0,0,2,4
2470,NA20287,0,0,2,1
2436,NA19713,0,0,2,3</code></pre><h4 id="Genetic-file(s)"><a class="docs-heading-anchor" href="#Genetic-file(s)">Genetic file(s)</a><a id="Genetic-file(s)-1"></a><a class="docs-heading-anchor-permalink" href="#Genetic-file(s)" title="Permalink"></a></h4><p>OrdinalGWAS supports PLINK, VCF, and BGEN Files.</p><ul><li><p>Genotype data is available as binary PLINK files.</p></li><li><p>OrdinalGWAS can use dosage or genotype data from VCF Files. </p></li><li><p>OrdinalGWAS can use dosage data from BGEN files.   </p></li></ul><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>By default, OrdinalGWAS assumes a set of PLINK files will be used. When using a VCF File, VCF file and type of data (dosage, genotype) must be specified by the <code>geneticformat</code> and <code>vcftype</code> options (as shown later). Similarly, BGEN must be specified as the <code>geneticformat</code> if using a BGEN file. </p></div></div><pre><code class="language-julia hljs">readdir(glob&quot;hapmap3.*&quot;, datadir)</code></pre><pre><code class="nohighlight hljs">4-element Vector{String}:
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.map&quot;</code></pre><p>In this example, there are 324 samples at 13,928 SNPs.</p><pre><code class="language-julia hljs">size(SnpArray(datadir * &quot;hapmap3.bed&quot;))</code></pre><pre><code class="nohighlight hljs">(324, 13928)</code></pre><p>Compressed PLINK and VCF files are supported. For example, if Plink files are <code>hapmap3.bed.gz</code>, <code>hapmap3.bim.gz</code> and <code>hapmap3.fam.gz</code>, the same command</p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;)</code></pre><p>still works. Check all supported compression format by</p><pre><code class="language-julia hljs">SnpArrays.ALLOWED_FORMAT</code></pre><pre><code class="nohighlight hljs">6-element Vector{String}:
 &quot;gz&quot;
 &quot;zlib&quot;
 &quot;zz&quot;
 &quot;xz&quot;
 &quot;zst&quot;
 &quot;bz2&quot;</code></pre><h3 id="Output-files"><a class="docs-heading-anchor" href="#Output-files">Output files</a><a id="Output-files-1"></a><a class="docs-heading-anchor-permalink" href="#Output-files" title="Permalink"></a></h3><p><code>ordinalgwas</code> outputs two files: <code>ordinalgwas.null.txt</code> and <code>ordinalgwas.pval.txt</code>. </p><ul><li><code>ordinalgwas.null.txt</code> lists the estimated null model (without SNPs). </li></ul><pre><code class="language-julia hljs">run(`cat ordinalgwas.null.txt`);</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><ul><li><code>ordinalgwas.pval.txt</code> tallies the SNPs and their pvalues. </li></ul><pre><code class="language-julia hljs">run(`head ordinalgwas.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,maf,hwepval,pval
1,554484,rs10458597,0.0,1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.4098763332666681,0.004565312839535881
1,967643,rs2710875,0.32407407407407407,4.076249100705747e-7,3.108283828553597e-5
1,1168108,rs11260566,0.19158878504672894,0.1285682279446898,1.2168672367654906e-5
1,1375074,rs1312568,0.441358024691358,2.5376019650614977e-19,0.008206860046174836
1,1588771,rs35154105,0.0,1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,0.5111981332534471
1,1990452,rs2678939,0.4537037037037037,5.07695957708431e-11,0.29972829571847304
1,2194615,rs7553178,0.22685185185185186,0.17056143157457776,0.17133312458049288</code></pre><p>Output file names can be changed by the <code>nullfile</code> and <code>pvalfile</code> keywords respectively. For example, </p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;,
    pvalfile=&quot;ordinalgwas.pval.txt.gz&quot;)</code></pre><p>will output the p-value file in compressed gz format.</p><h3 id="Subsamples"><a class="docs-heading-anchor" href="#Subsamples">Subsamples</a><a id="Subsamples-1"></a><a class="docs-heading-anchor-permalink" href="#Subsamples" title="Permalink"></a></h3><p>Use the keyword <code>covrowinds</code> to specify selected samples in the covarite file. Use the keyword <code>geneticrowinds</code> to specify selected samples in the Plink bed file or VCF File. For example, to use the first 300 samples in both covariate and bed file:</p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), covfile, geneticfile, 
    covrowinds=1:300, geneticrowinds=1:300)</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Users should always make sure that the selected samples in covariate file match exactly those in bed file. </p></div></div><h3 id="Input-non-genetic-data-as-DataFrame"><a class="docs-heading-anchor" href="#Input-non-genetic-data-as-DataFrame">Input non-genetic data as DataFrame</a><a id="Input-non-genetic-data-as-DataFrame-1"></a><a class="docs-heading-anchor-permalink" href="#Input-non-genetic-data-as-DataFrame" title="Permalink"></a></h3><p>Internally <code>ordinalgwas</code> parses the covariate file as a DataFrame by <code>CSV.read(covfile)</code>. For covariate file of other formats, users can parse it as a DataFrame and then input the DataFrame to <code>ordinalgwas</code> directly.</p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), df, geneticfile)</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Users should always make sure that individuals in covariate file or DataFrame match those in Plink fam file/VCF File/BGEN file. </p></div></div><p>For example, following code checks that the first 2 columns of the <code>covariate.txt</code> file match the first 2 columns of the <code>hapmap3.fam</code> file exactly.</p><pre><code class="language-julia hljs">covdf = CSV.read(datadir * &quot;covariate.txt&quot;, DataFrame)
plkfam = CSV.read(datadir * &quot;hapmap3.fam&quot;, header=0, delim=&#39; &#39;,  DataFrame)
all(covdf[!, 1] .== plkfam[!, 1]) &amp;&amp; all(covdf[!, 2] .== plkfam[!, 2])</code></pre><pre><code class="nohighlight hljs">true</code></pre><h3 id="Timing"><a class="docs-heading-anchor" href="#Timing">Timing</a><a id="Timing-1"></a><a class="docs-heading-anchor-permalink" href="#Timing" title="Permalink"></a></h3><p>For this moderate-sized data set, <code>ordinalgwas</code> takes around 0.2 seconds.</p><pre><code class="language-julia hljs">@btime(ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;));</code></pre><pre><code class="nohighlight hljs">  201.823 ms (500794 allocations: 49.17 MiB)</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;ordinalgwas.null.txt&quot;, force=true)
rm(&quot;ordinalgwas.pval.txt&quot;, force=true)</code></pre><h2 id="VCF-Formatted-Files"><a class="docs-heading-anchor" href="#VCF-Formatted-Files">VCF Formatted Files</a><a id="VCF-Formatted-Files-1"></a><a class="docs-heading-anchor-permalink" href="#VCF-Formatted-Files" title="Permalink"></a></h2><p>By default, OrdinalGWAS.jl will assume you are using PLINK files. It also supports VCF (and BGEN) Files. <code>vcf.gz</code> files are supported as well. To use vcf files in any of the analysis options detailed in this documentation, you simply need to add two keyword options to the <code>ordinalgwas</code> function:</p><ul><li><code>geneticformat</code>: Choices are &quot;VCF&quot; or &quot;PLINK&quot;. If you are using a VCF file, use <code>geneticformat = &quot;VCF&quot;</code>.</li><li><code>vcftype</code>: Choices are :GT (for genotypes) or :DS (for dosages). This tells OrdinalGWAS which type of data to extract from the VCF file.</li></ul><p>Using a VCF File does not output minor allele frequency or hardy weinberg equillibrium p-values for each SNP tested since they may be dosages and MAF/HWE calculations via VCFTools does not currently support dosage data. </p><p>The following shows how to run an analysis with a VCF file using the dosage information. </p><pre><code class="language-julia hljs">ordinalgwas(@formula(y ~ sex), datadir * &quot;vcf_example.csv&quot;, 
    datadir * &quot;vcf_test&quot;; geneticformat = &quot;VCF&quot;, vcftype = :DS)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

y ~ sex

Coefficients:
───────────────────────────────────────────────────────
               Estimate  Std.Error    t value  Pr(&gt;|t|)
───────────────────────────────────────────────────────
intercept1|2  -1.10106    0.205365  -5.36147     &lt;1e-06
intercept2|3   0.370894   0.188822   1.96425     0.0510
intercept3|4   1.74736    0.232756   7.50726     &lt;1e-11
sex            0.22796    0.262237   0.869289    0.3858
───────────────────────────────────────────────────────</code></pre><pre><code class="language-julia hljs">run(`head ordinalgwas.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,pval
22,20000086,rs138720731,0.6435068072069544
22,20000146,rs73387790,1.0
22,20000199,rs183293480,0.9378258278500581
22,20000291,rs185807825,0.21907288710091172
22,20000428,rs55902548,0.002790402446884929
22,20000683,rs142720028,1.0
22,20000771,rs114690707,0.23034910158749672
22,20000793,rs189842693,0.15612228776953083
22,20000810,rs147349046,0.1741386270145462</code></pre><h2 id="BGEN-Formatted-Files"><a class="docs-heading-anchor" href="#BGEN-Formatted-Files">BGEN Formatted Files</a><a id="BGEN-Formatted-Files-1"></a><a class="docs-heading-anchor-permalink" href="#BGEN-Formatted-Files" title="Permalink"></a></h2><p>By default, OrdinalGWAS.jl will assume you are using PLINK files. It also supports BGEN Files. To use BGEN files in any of the analysis options detailed in this documentation, you simply need to add the following keyword option to the <code>ordinalgwas</code> function:</p><ul><li><code>geneticformat</code>: Choices are &quot;VCF&quot; or &quot;PLINK&quot; or &quot;BGEN&quot;. If you are using a BGEN file, use <code>geneticformat = &quot;BGEN&quot;</code>.</li></ul><p>Some features, such as SNP-set analyses, are only available currently when there&#39;s an indexing <code>.bgi</code> file available. </p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>BGEN files can contain an optional index file (<code>.bgi</code> file) that allows the variants to be specified in order of position. OrdinalGWAS will automatically look for a file in the same directory as the <code>BGENFILENAME</code> with the name <code>BGENFILENAME.bgi</code>. The BGEN file is read either in the <code>.bgi</code> file order if <code>BGENFILENAME.bgi</code> is supplied in the same directory as the BGEN file, otherwise it will use the order in the BGEN file. This is important in analyses specifying <code>snpinds</code> as well as annotation groupings. You must make sure this matches the way the BGEN file will be read in. </p></div></div><p>The following shows how to run an analysis with a BGEN file using the dosage information. </p><pre><code class="language-julia hljs">ordinalgwas(@formula(y ~ sex), datadir * &quot;bgen_ex.csv&quot;, 
    datadir * &quot;bgen_test&quot;; geneticformat = &quot;BGEN&quot;)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

y ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.08828    0.128371  -8.47759    &lt;1e-15
intercept2|3   0.434834   0.118541   3.66821    0.0003
intercept3|4   1.80196    0.145367  12.396      &lt;1e-30
sex            0.461387   0.162676   2.83623    0.0048
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia hljs">run(`head ordinalgwas.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,varid,maf,hwepval,infoscore,pval
01,1001,RSID_101,SNPID_101,0.4169765,0.8627403,0.9846387832074154,0.12449779722495212
01,2000,RSID_2,SNPID_2,0.19750981,0.1921813,9.0,0.0005572706473664106
01,2001,RSID_102,SNPID_102,0.19766665,0.18440013,0.72730855295163,0.0004996980174197089
01,3000,RSID_3,SNPID_3,0.48339605,0.9653543,0.955355323674357,0.5866179557677308
01,3001,RSID_103,SNPID_103,0.4833961,0.9653536,0.9553553468765227,0.5866179209956521
01,4000,RSID_4,SNPID_4,0.21670982,0.37192723,0.9917677420119885,0.3522822949083395
01,4001,RSID_104,SNPID_104,0.2167098,0.37192774,0.99176792776617,0.35228235843382233
01,5000,RSID_5,SNPID_5,0.38808233,0.5870128,0.9682577646241436,0.8349949433297409
01,5001,RSID_105,SNPID_105,0.3880863,0.5867037,0.9682302221772928,0.8347201604324502</code></pre><h2 id="Link-functions"><a class="docs-heading-anchor" href="#Link-functions">Link functions</a><a id="Link-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Link-functions" title="Permalink"></a></h2><p>The <code>link</code> keyword argument of <code>ordinalgwas</code> can take value:  </p><ul><li><code>LogitLink()</code>, proportional odds model (default),  </li><li><code>ProbitLink()</code>, ordred Probit model,  </li><li><code>CloglogLink()</code>, proportional hazards model, or </li><li><code>CauchyLink()</code>.</li></ul><p>For example, to perform GWAS using the ordred Probit model</p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    link=ProbitLink(), nullfile=&quot;opm.null.txt&quot;, pvalfile=&quot;opm.pval.txt&quot;)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, ProbitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -0.866156   0.210677  -4.11129    &lt;1e-04
intercept2|3  -0.359878   0.205817  -1.74854    0.0813
intercept3|4   0.247054   0.205382   1.2029     0.2299
sex            0.251058   0.128225   1.95795    0.0511
──────────────────────────────────────────────────────</code></pre><p>The estimates in null model and p-values are slightly different from those in proportional odds moodel.</p><pre><code class="language-julia hljs">run(`head opm.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,maf,hwepval,pval
1,554484,rs10458597,0.0,1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.4098763332666681,0.010076916742307648
1,967643,rs2710875,0.32407407407407407,4.076249100705747e-7,2.62725649418621e-5
1,1168108,rs11260566,0.19158878504672894,0.1285682279446898,1.0897484851087357e-5
1,1375074,rs1312568,0.441358024691358,2.5376019650614977e-19,0.0051028839904383545
1,1588771,rs35154105,0.0,1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,0.48653776297937934
1,1990452,rs2678939,0.4537037037037037,5.07695957708431e-11,0.33231290090456195
1,2194615,rs7553178,0.22685185185185186,0.17056143157457776,0.25915513977199217</code></pre><pre><code class="language-julia hljs">rm(&quot;opm.null.txt&quot;, force=true)
rm(&quot;opm.pval.txt&quot;, force=true)</code></pre><h2 id="SNP-models"><a class="docs-heading-anchor" href="#SNP-models">SNP models</a><a id="SNP-models-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-models" title="Permalink"></a></h2><p>Genotypes are translated into numeric values according to different genetic model, which is specified by the <code>snpmodel</code> keyword. Default is <code>ADDITIVE_MODEL</code>.</p><table><tr><th style="text-align: center">Genotype</th><th style="text-align: center"><code>SnpArray</code></th><th style="text-align: center"><code>ADDITIVE_MODEL</code></th><th style="text-align: center"><code>DOMINANT_MODEL</code></th><th style="text-align: center"><code>RECESSIVE_MODEL</code></th></tr><tr><td style="text-align: center">A1,A1</td><td style="text-align: center">0x00</td><td style="text-align: center">0</td><td style="text-align: center">0</td><td style="text-align: center">0</td></tr><tr><td style="text-align: center">missing</td><td style="text-align: center">0x01</td><td style="text-align: center">NaN</td><td style="text-align: center">NaN</td><td style="text-align: center">NaN</td></tr><tr><td style="text-align: center">A1,A2</td><td style="text-align: center">0x02</td><td style="text-align: center">1</td><td style="text-align: center">1</td><td style="text-align: center">0</td></tr><tr><td style="text-align: center">A2,A2</td><td style="text-align: center">0x03</td><td style="text-align: center">2</td><td style="text-align: center">1</td><td style="text-align: center">1</td></tr></table><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>ordinalgwas</code> imputes missing genotypes according to minor allele frequencies. </p></div></div><p>Users are advised to impute genotypes using more sophiscated methods before GWAS.</p><h2 id="SNP-and/or-sample-masks"><a class="docs-heading-anchor" href="#SNP-and/or-sample-masks">SNP and/or sample masks</a><a id="SNP-and/or-sample-masks-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-and/or-sample-masks" title="Permalink"></a></h2><p>In practice, we often perform GWAS on selected SNPs and/or selected samples. They can be specified by the <code>snpinds</code>, <code>covrowinds</code> and <code>geneticrowinds</code> keywords of <code>ordinalgwas</code> function. </p><p>For example, to perform GWAS on SNPs with minor allele frequency (MAF) above 0.05</p><pre><code class="language-julia hljs"># create SNP mask
snpinds = maf(SnpArray(datadir * &quot;hapmap3.bed&quot;)) .≥ 0.05
# GWAS on selected SNPs
@time ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    snpinds=snpinds, nullfile=&quot;commonvariant.null.txt&quot;, pvalfile=&quot;commonvariant.pval.txt&quot;)</code></pre><pre><code class="nohighlight hljs">  1.578096 seconds (3.07 M allocations: 198.051 MiB, 8.05% gc time, 86.59% compilation time)





StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia hljs">run(`head commonvariant.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,maf,hwepval,pval
1,758311,rs12562034,0.07763975155279501,0.4098763332666681,0.004565312839535881
1,967643,rs2710875,0.32407407407407407,4.076249100705747e-7,3.108283828553597e-5
1,1168108,rs11260566,0.19158878504672894,0.1285682279446898,1.2168672367654906e-5
1,1375074,rs1312568,0.441358024691358,2.5376019650614977e-19,0.008206860046174836
1,1990452,rs2678939,0.4537037037037037,5.07695957708431e-11,0.29972829571847304
1,2194615,rs7553178,0.22685185185185186,0.17056143157457776,0.17133312458049288
1,2396747,rs13376356,0.1448598130841121,0.9053079215078139,0.5320416198875104
1,2823603,rs1563468,0.4830246913580247,4.23065537243926e-9,0.22519139178356798
1,3025087,rs6690373,0.2538699690402477,9.238641887192776e-8,0.7018469417717436</code></pre><pre><code class="language-julia hljs"># extra header line in commonvariant.pval.txt
countlines(&quot;commonvariant.pval.txt&quot;), count(snpinds)</code></pre><pre><code class="nohighlight hljs">(12086, 12085)</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;commonvariant.null.txt&quot;, force=true)
rm(&quot;commonvariant.pval.txt&quot;, force=true)</code></pre><p><code>covrowinds</code> specify the samples in the covariate file and <code>geneticrowinds</code> for PLINK or VCF File. User should be particularly careful when using these two keyword. Selected rows in SnpArray should exactly match the samples in the null model. Otherwise the results are meaningless.</p><h2 id="Likelihood-ratio-test-(LRT)"><a class="docs-heading-anchor" href="#Likelihood-ratio-test-(LRT)">Likelihood ratio test (LRT)</a><a id="Likelihood-ratio-test-(LRT)-1"></a><a class="docs-heading-anchor-permalink" href="#Likelihood-ratio-test-(LRT)" title="Permalink"></a></h2><p>By default, <code>ordinalgwas</code> calculates p-value for each SNP using score test. Score test is fast because it doesn&#39;t require fitting alternative model for each SNP. User can request likelihood ratio test (LRT) using keyword <code>test=:lrt</code>. LRT is much slower but may be more powerful than score test.</p><p>Because LRT fits the alternative model for each SNP, we also output the standard error and estimated effect size of the SNP. </p><pre><code class="language-julia hljs">@time ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    test=:LRT, nullfile=&quot;lrt.null.txt&quot;, pvalfile=&quot;lrt.pval.txt&quot;)</code></pre><pre><code class="nohighlight hljs"> 21.716944 seconds (5.51 M allocations: 1.911 GiB, 1.72% gc time, 1.84% compilation time)





StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><p>Note the extra <code>effect</code> and <code>stder</code> columns in pvalfile, which is the effect size (regression coefficient) and standard errors for each SNP. <code>-1</code> is reported when there is no variation in the SNP (0 MAF).  </p><pre><code class="language-julia hljs">run(`head lrt.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,maf,hwepval,effect,stder,pval
1,554484,rs10458597,0.0,1.0,0.0,-1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.4098763332666681,-1.0057833719544336,0.34080917686608586,0.0019185836579804134
1,967643,rs2710875,0.32407407407407407,4.076249100705747e-7,-0.6488560566291872,0.15659100275284915,1.805050556976133e-5
1,1168108,rs11260566,0.19158878504672894,0.1285682279446898,-0.9157225669357891,0.2177988861470356,5.873384712686268e-6
1,1375074,rs1312568,0.441358024691358,2.5376019650614977e-19,-0.3318136652577256,0.12697404813031218,0.008081022577833346
1,1588771,rs35154105,0.0,1.0,0.0,-1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,-0.733802638869998,1.2495287571216083,0.5169027130144458
1,1990452,rs2678939,0.4537037037037037,5.07695957708431e-11,-0.13586499231819757,0.13123472232496708,0.29946402200912603
1,2194615,rs7553178,0.22685185185185186,0.17056143157457776,-0.251207564044013,0.17872139152979566,0.1615106909443865</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;lrt.pval.txt&quot;, force=true)
rm(&quot;lrt.null.txt&quot;, force=true)</code></pre><p>In this example, GWAS by score test takes less than 0.2 second, while GWAS by LRT takes over 20 seconds. Over 100 fold difference in run time. </p><h2 id="Score-test-for-screening,-LRT-for-power"><a class="docs-heading-anchor" href="#Score-test-for-screening,-LRT-for-power">Score test for screening, LRT for power</a><a id="Score-test-for-screening,-LRT-for-power-1"></a><a class="docs-heading-anchor-permalink" href="#Score-test-for-screening,-LRT-for-power" title="Permalink"></a></h2><p>For large data sets, a practical solution is to perform the score test first across all SNPs, then re-do LRT for the most promising SNPs according to score test p-values.</p><p><strong>Step 1</strong>: Perform score test GWAS, results in <code>score.pval.txt</code>.</p><pre><code class="language-julia hljs">@time ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    test=:score, pvalfile=&quot;score.pval.txt&quot;);</code></pre><pre><code class="nohighlight hljs">  0.328529 seconds (500.80 k allocations: 49.171 MiB)</code></pre><pre><code class="language-julia hljs">run(`head score.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,maf,hwepval,pval
1,554484,rs10458597,0.0,1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.4098763332666681,0.004565312839535881
1,967643,rs2710875,0.32407407407407407,4.076249100705747e-7,3.108283828553597e-5
1,1168108,rs11260566,0.19158878504672894,0.1285682279446898,1.2168672367654906e-5
1,1375074,rs1312568,0.441358024691358,2.5376019650614977e-19,0.008206860046174836
1,1588771,rs35154105,0.0,1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,0.5111981332534471
1,1990452,rs2678939,0.4537037037037037,5.07695957708431e-11,0.29972829571847304
1,2194615,rs7553178,0.22685185185185186,0.17056143157457776,0.17133312458049288</code></pre><p><strong>Step 2</strong>: Sort score test p-values and find top 10 SNPs.</p><pre><code class="language-julia hljs">scorepvals = CSV.read(&quot;score.pval.txt&quot;, DataFrame)[!, end] # p-values in last column
tophits = sortperm(scorepvals)[1:10] # indices of 10 SNPs with smallest p-values
scorepvals[tophits] # smallest 10 p-values</code></pre><pre><code class="nohighlight hljs">10-element Vector{Float64}:
 1.3080149099173788e-6
 6.536722765046125e-6
 9.664742185663502e-6
 1.2168672367654906e-5
 1.8027460018322353e-5
 2.0989542284210004e-5
 2.6844521269932935e-5
 3.108283828553597e-5
 4.101091287507263e-5
 4.2966265138340985e-5</code></pre><p><strong>Step 3</strong>: Re-do LRT on top hits.</p><pre><code class="language-julia hljs">@time ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    snpinds=tophits, test=:LRT, pvalfile=&quot;lrt.pval.txt&quot;);</code></pre><pre><code class="nohighlight hljs">  1.321790 seconds (1.91 M allocations: 114.008 MiB, 3.17% gc time, 95.63% compilation time)</code></pre><pre><code class="language-julia hljs">run(`cat lrt.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,maf,hwepval,effect,stder,pval
1,967643,rs2710875,0.32407407407407407,4.076249100705747e-7,-0.6488560566291872,0.15659100275284915,1.805050556976133e-5
1,1168108,rs11260566,0.19158878504672894,0.1285682279446898,-0.9157225669357891,0.2177988861470356,5.873384712686268e-6
3,36821790,rs4678553,0.23456790123456794,0.1094668216324497,0.7424952268973514,0.17055414498937366,1.1303825016262592e-5
4,11017683,rs16881446,0.27554179566563464,0.8942746118760274,-0.7870581482955514,0.18729099511224528,1.1105427468799613e-5
5,3739190,rs12521166,0.0679012345679012,0.18613647746093887,1.1468852997925314,0.2809307861547482,4.7812882296576845e-5
6,7574576,rs1885466,0.17746913580246915,0.7620687178987191,0.8750621092263018,0.19417734621377952,7.272346896740631e-6
6,52474721,rs2073183,0.1826625386996904,0.5077765730476698,0.7790794914858653,0.19764574359737597,5.069394513904594e-5
7,41152376,rs28880,0.3379629629629629,0.8052368892744096,-0.8146339024453504,0.17471459882610024,9.180126530294943e-7
7,84223996,rs4128623,0.07870370370370372,0.0218347173467568,1.0022229316338562,0.25534365673199283,6.587895464657107e-5
23,121048059,rs1937165,0.4380804953560371,3.959609737265113e-16,0.5392313636256605,0.1280959604207797,1.975464385552384e-5</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;ordinalgwas.null.txt&quot;, force=true)
rm(&quot;score.pval.txt&quot;, force=true)
rm(&quot;lrt.pval.txt&quot;, force=true)</code></pre><h2 id="GxE-or-other-interactions"><a class="docs-heading-anchor" href="#GxE-or-other-interactions">GxE or other interactions</a><a id="GxE-or-other-interactions-1"></a><a class="docs-heading-anchor-permalink" href="#GxE-or-other-interactions" title="Permalink"></a></h2><h3 id="Testing-jointly-G-GxE"><a class="docs-heading-anchor" href="#Testing-jointly-G-GxE">Testing jointly G + GxE</a><a id="Testing-jointly-G-GxE-1"></a><a class="docs-heading-anchor-permalink" href="#Testing-jointly-G-GxE" title="Permalink"></a></h3><p>In many applications, we want to test SNP effect and/or its interaction with other terms. <code>testformula</code> keyword specifies the test unit <strong>besides</strong> the covariates in <code>nullformula</code>. </p><p>In following example, keyword <code>testformula=@formula(trait ~ snp + snp &amp; sex)</code> instructs <code>ordinalgwas</code> to test joint effect of <code>snp</code> and <code>snp &amp; sex</code> interaction.</p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    pvalfile=&quot;GxE.pval.txt&quot;, testformula=@formula(trait ~ snp + snp &amp; sex));</code></pre><pre><code class="language-julia hljs">run(`head GxE.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,maf,hwepval,pval
1,554484,rs10458597,0.0,1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.4098763332666681,0.017446010412245565
1,967643,rs2710875,0.32407407407407407,4.076249100705747e-7,0.0001667073239488232
1,1168108,rs11260566,0.19158878504672894,0.1285682279446898,4.7637624578926465e-5
1,1375074,rs1312568,0.441358024691358,2.5376019650614977e-19,0.029138471242991387
1,1588771,rs35154105,0.0,1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,0.29643631147339605
1,1990452,rs2678939,0.4537037037037037,5.07695957708431e-11,0.3792458047934272
1,2194615,rs7553178,0.22685185185185186,0.17056143157457776,0.32558226993239</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;ordinalgwas.null.txt&quot;)
rm(&quot;GxE.pval.txt&quot;)</code></pre><h3 id="Testing-only-GxE-interaction-term"><a class="docs-heading-anchor" href="#Testing-only-GxE-interaction-term">Testing only GxE interaction term</a><a id="Testing-only-GxE-interaction-term-1"></a><a class="docs-heading-anchor-permalink" href="#Testing-only-GxE-interaction-term" title="Permalink"></a></h3><p>For some applications, the user may want to simply test the GxE interaction effect. This requires fitting the SNP in the null model and is much slower, but the command <code>ordinalgwas()</code> with keyword <code>analysistype = &quot;gxe&quot;</code> can be used test the interaction effect. The environmental variable must be specified in the command using the keyword argument <code>e</code>, either as a symbol, such as <code>:age</code> or as a string <code>&quot;age&quot;</code>. </p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;; analysistype = &quot;gxe&quot;,
    e = :sex, pvalfile = &quot;gxe_snp.pval.txt&quot;, snpinds=1:5, test=:score)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia hljs">run(`head gxe_snp.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr,pos,snpid,maf,hwepval,snpeffectnull,pval
1,554484,rs10458597,0.0,1.0,0.0,1.0
1,758311,rs12562034,0.07763975155279501,0.4098763332666681,-1.0057833719544336,0.6377422425980503
1,967643,rs2710875,0.32407407407407407,4.076249100705747e-7,-0.6488560566291872,0.9667114197304937
1,1168108,rs11260566,0.19158878504672894,0.1285682279446898,-0.9157225669357891,0.263526746940882
1,1375074,rs1312568,0.441358024691358,2.5376019650614977e-19,-0.3318136652577256,0.7811133315583378</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;gxe_snp.pval.txt&quot;, force=true)</code></pre><h2 id="SNP-set-testing"><a class="docs-heading-anchor" href="#SNP-set-testing">SNP-set testing</a><a id="SNP-set-testing-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-set-testing" title="Permalink"></a></h2><p>In many applications, we want to test a SNP-set. The function with keyword <code>analysistype = &quot;snpset&quot;</code> can be used to do this. To specify the type of snpset test, use the <code>snpset</code> keyword argument. </p><p>The snpset can be specified as either:</p><ul><li>a window (test every X snps) =&gt; <code>snpset = X</code></li><li>an annotated file.  This requires <code>snpset = filename</code> where filename is an input file, with no header and two columns separated by a space. The first column must contain the snpset ID and the second column must contain the snpid&#39;s (rsid) identical to the bimfile, or in the case of a BGEN format, the order the data will be read (see BGEN above for more details).</li><li>a joint test on only a specific set of snps. <code>snpset = AbstractVector</code> where the vector specifies the snps you want to perform one joint snpset test for. The vector can either be a vector of integers where the elements are the indicies of the SNPs to test, a vector of booleans, where true represents that you want to select that SNP index, or a range indicating the indicies of the SNPs to test. </li></ul><p>In following example, we perform a SNP-set test on the 50th to 55th snps. </p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;; 
    analysistype = &quot;snpset&quot;, pvalfile = &quot;snpset.pval.txt&quot;, snpset = 50:55)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia hljs">run(`head snpset.pval.txt`);</code></pre><pre><code class="nohighlight hljs">The joint pvalue of snps indexed at 50:55 is 0.3647126536663968</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;ordinalgwas.null.txt&quot;, force=true)</code></pre><p>In the following example we run a SNP-set test on the annotated SNP-set file.</p><pre><code class="language-julia hljs">run(`head $datadir/hapmap_snpsetfile.txt`);</code></pre><pre><code class="nohighlight hljs">gene1 rs10458597
gene1 rs12562034
gene1 rs2710875
gene1 rs11260566
gene1 rs1312568
gene1 rs35154105
gene1 rs16824508
gene1 rs2678939
gene1 rs7553178
gene1 rs13376356</code></pre><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;;
    analysistype = &quot;snpset&quot;, pvalfile = &quot;snpset.pval.txt&quot;, 
    snpset = datadir * &quot;/hapmap_snpsetfile.txt&quot;)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia hljs">run(`head snpset.pval.txt`);</code></pre><pre><code class="nohighlight hljs">snpsetid,nsnps,pval
gene1,93,1.7213394698790313e-5
gene2,93,0.03692497684163047
gene3,93,0.7478549371392569
gene4,92,0.02765079822347653
gene5,93,0.6119581594573154
gene6,93,0.029184642230066303
gene7,93,0.39160073488514746
gene8,93,0.12539574109853385
gene9,93,0.7085635621607591</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;ordinalgwas.null.txt&quot;, force=true)</code></pre><p>In the following example we run a SNP-set test on every 15 SNPs.</p><pre><code class="language-julia hljs">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;;
    analysistype = &quot;snpset&quot;, pvalfile = &quot;snpset.pval.txt&quot;, snpset=15)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia hljs">run(`head snpset.pval.txt`);</code></pre><pre><code class="nohighlight hljs">startchr,startpos,startsnpid,endchr,endpos,endsnpid,pval
1,554484,rs10458597,1,3431124,rs12093117,1.9394189465428366e-13
1,3633945,rs10910017,1,6514524,rs932112,0.11077869162800988
1,6715827,rs441515,1,9534606,rs4926480,0.2742450817958276
1,9737551,rs12047054,1,12559747,rs4845907,0.49346113650467543
1,12760427,rs848577,1,16021797,rs6679870,0.15447358658256338
1,16228774,rs1972359,1,19100349,rs9426794,0.1544232917023417
1,19301516,rs4912046,1,22122176,rs9426785,0.4749873349462694
1,22323074,rs2235529,1,25166528,rs4648890,0.41246096621461137
1,25368553,rs7527379,1,28208168,rs12140070,0.16458135278661626</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;ordinalgwas.null.txt&quot;, force=true)</code></pre><h2 id="Plotting-Results"><a class="docs-heading-anchor" href="#Plotting-Results">Plotting Results</a><a id="Plotting-Results-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting-Results" title="Permalink"></a></h2><p>To plot the GWAS results, we recommend using the <a href="https://openmendel.github.io/MendelPlots.jl/latest/">MendelPlots.jl package</a>.</p><h2 id="Docker"><a class="docs-heading-anchor" href="#Docker">Docker</a><a id="Docker-1"></a><a class="docs-heading-anchor-permalink" href="#Docker" title="Permalink"></a></h2><p>For ease of using OrdinalGWAS, we provide a Dockerfile so users don&#39;t need to install Julia and required packages. Only Docker app needs to be installed in order to run analysis. Following is tested on Docker 2.0.0.0-mac78.</p><p><strong>Step 1</strong>: Create a Dockerfile with content <a href="https://raw.githubusercontent.com/OpenMendel/OrdinalGWAS.jl/master/docker/Dockerfile">here</a>, or, if the bash command <code>wget</code> is available, obtain Dockerfile by</p><pre><code class="language-bash hljs"># on command line
wget https://raw.githubusercontent.com/OpenMendel/OrdinalGWAS.jl/master/docker/Dockerfile</code></pre><p><strong>Step 2</strong>: Build a docker image called <code>ordinalgwas-app</code>, assuming that the Dockerfile is located in the <code>../docker</code> folder. Building the image for the first time can take up to 10 minutes; but it only needs to be done once.</p><pre><code class="language-bash hljs"># on command line
docker build -t ordinalgwas-app ../docker/</code></pre><p><strong>Step 3</strong>: Suppose data files are located at <code>/path/to/data</code> folder, run analysis by</p><pre><code class="language-bash hljs"># on command line
docker run -v /path/to/data:/data -t ordinalgwas-app julia -e &#39;using OrdinalGWAS; ordinalgwas(@formula(trait ~ sex), &quot;/data/covariate.txt&quot;, &quot;/data/hapmap3&quot;, nullfile=&quot;/data/ordinalgwas.null.txt&quot;, pvalfile=&quot;/data/ordinalgwas&quot;);&#39;</code></pre><p>Here  </p><ul><li><code>-t ordinalgwas-app</code> creates a container using the <code>ordinalgwas-app</code> image build in step 2.  </li><li><code>-v /path/to/data:/data</code> maps the <code>/path/to/data</code> folder on host machine to the <code>/data</code> folder within the container. </li><li><code>julia -e &#39;using OrdinalGWAS; ordinalgwas(@formula(trait ~ sex), &quot;/data/covariate.txt&quot;, &quot;/data/hapmap3&quot;, nullfile=&quot;/data/ordinalgwas.null.txt&quot;, pvalfile=&quot;/data/ordinalgwas&quot;);</code> calls Julia and runs <code>ordinalgwas</code> function. </li></ul><p>The output files are written in <code>/path/to/data</code> directory.</p><h2 id="Multiple-Plink-file-sets"><a class="docs-heading-anchor" href="#Multiple-Plink-file-sets">Multiple Plink file sets</a><a id="Multiple-Plink-file-sets-1"></a><a class="docs-heading-anchor-permalink" href="#Multiple-Plink-file-sets" title="Permalink"></a></h2><p>In large scale studies, genotypes data are split into multiple Plink files, e.g., by chromosome. Then GWAS analysis can be done in parallel. This can be achieved by two steps.</p><p>Let&#39;s first create demo data by splitting hapmap3 according to chromosome:</p><pre><code class="language-julia hljs"># split example hapmap3 data according to chromosome
SnpArrays.split_plink(datadir * &quot;hapmap3&quot;, :chromosome; prefix=datadir * &quot;hapmap3.chr.&quot;)
readdir(glob&quot;hapmap3.chr.*&quot;, datadir)</code></pre><pre><code class="nohighlight hljs">75-element Vector{String}:
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.1.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.1.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.1.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.10.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.10.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.10.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.11.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.11.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.11.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.12.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.12.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.12.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.13.bed&quot;
 ⋮
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.6.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.6.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.6.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.7.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.7.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.7.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.8.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.8.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.8.fam&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.9.bed&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.9.bim&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.9.fam&quot;</code></pre><p>Step 1: Fit the null model. Setting third argument <code>geneticfile</code> to <code>nothing</code> instructs <code>ordinalgwas</code> function to fit the null model only.</p><pre><code class="language-julia hljs">nm = ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, nothing)</code></pre><pre><code class="nohighlight hljs">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64, Float64, LogitLink}, Matrix{Float64}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-04
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><p>Step 2: GWAS for each chromosome.</p><pre><code class="language-julia hljs"># this part can be submitted as separate jobs
for chr in 1:23
    plinkfile = datadir * &quot;hapmap3.chr.&quot; * string(chr)
    pvalfile = plinkfile * &quot;.pval.txt&quot; 
    ordinalgwas(nm, plinkfile, pvalfile=pvalfile)
end</code></pre><pre><code class="language-julia hljs"># show the result files
readdir(glob&quot;*.pval.txt&quot;, datadir)</code></pre><pre><code class="nohighlight hljs">23-element Vector{String}:
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.1.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.10.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.11.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.12.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.13.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.14.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.15.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.16.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.17.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.18.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.19.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.2.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.20.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.21.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.22.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.23.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.3.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.4.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.5.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.6.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.7.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.8.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/OrdinalGWAS/data/hapmap3.chr.9.pval.txt&quot;</code></pre><p>In the rare situations where the multiple sets of Plink files lack the <code>fam</code> file or the corresponding bed and bim files have different filenames, users can explicitly supply bed filename, bim file name, and number of individuals. Replace Step 2 by </p><p>Step 2&#39;: GWAS for each chromosome.</p><pre><code class="language-julia hljs"># this part can be submitted as separate jobs
for chr in 1:23
    bedfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.bed&quot;
    bimfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.bim&quot;
    pvalfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.pval.txt&quot;
    ordinalgwas(nm, bedfile, bimfile, 324; pvalfile=pvalfile)
end</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;ordinalgwas.null.txt&quot;, force=true)
isfile(datadir * &quot;fittednullmodel.jld2&quot;) &amp;&amp; rm(datadir * &quot;fittednullmodel.jld2&quot;)
for chr in 1:23
    pvalfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.pval.txt&quot;
    rm(pvalfile, force=true)
end
for chr in 1:26
    plinkfile = datadir * &quot;hapmap3.chr.&quot; * string(chr)
    rm(plinkfile * &quot;.bed&quot;, force=true)
    rm(plinkfile * &quot;.fam&quot;, force=true)
    rm(plinkfile * &quot;.bim&quot;, force=true)
end</code></pre><h2 id="Multiple-Plink-file-sets-on-cluster"><a class="docs-heading-anchor" href="#Multiple-Plink-file-sets-on-cluster">Multiple Plink file sets on cluster</a><a id="Multiple-Plink-file-sets-on-cluster-1"></a><a class="docs-heading-anchor-permalink" href="#Multiple-Plink-file-sets-on-cluster" title="Permalink"></a></h2><p>We provide two scripts that successfully run on UCLA&#39;s Hoffman2 cluster using Julia v1.0.1 and PBS job schedulaer (<code>qsub</code>).</p><ul><li>The first script <a href="https://raw.githubusercontent.com/OpenMendel/OrdinalGWAS.jl/master/docs/cluster_preparedata.jl"><code>cluster_preparedata.jl</code></a> creates a demo data set in current folder. Run</li></ul><pre><code class="language-julia hljs">julia cluster_preparedata.jl</code></pre><p>on head node.</p><pre><code class="language-julia hljs">run(`cat $datadir/../docs/cluster_preparedata.jl`);</code></pre><pre><code class="nohighlight hljs">#!/usr/local/bin/julia
#
# This script prepares a data set in current folder. 
# For each of chromosome 1-23, there is a set gzipped Plink files:
# hapmap3.chr.1.bed.gz, hapmap3.chr.1.bim.gz, hapmap3.chr.1.fam.gz
# hapmap3.chr.2.bed.gz, hapmap3.chr.2.bim.gz, hapmap3.chr.2.fam.gz
# ...
# hapmap3.chr.23.bed.gz, hapmap3.chr.23.bim.gz, hapmap3.chr.23.fam.gz
# There is also a csv file &quot;covariate.txt&quot; that contains trait and covariates.
#

# install and load Julia packages
using Pkg
if haskey(Pkg.installed(), &quot;SnpArrays&quot;)
    Pkg.update(&quot;SnpArrays&quot;)
else
    Pkg.add(PackageSpec(url=&quot;https://github.com/OpenMendel/SnpArrays.jl.git&quot;))
end
if haskey(Pkg.installed(), &quot;OrdinalMultinomialModels&quot;)
    Pkg.update(&quot;OrdinalMultinomialModels&quot;)
else
    Pkg.add(PackageSpec(url=&quot;https://github.com/OpenMendel/OrdinalMultinomialModels.jl.git&quot;))
end
if haskey(Pkg.installed(), &quot;OrdinalGWAS&quot;)
    Pkg.update(&quot;OrdinalGWAS&quot;)
else
    Pkg.add(PackageSpec(url=&quot;https://github.com/OpenMendel/OrdinalGWAS.jl.git&quot;))
end
using OrdinalMultinomialModels, OrdinalGWAS, SnpArrays

# split hapmap3 data according to chromosome
datadir = normpath(joinpath(dirname(pathof(OrdinalGWAS)), &quot;../data/&quot;))
SnpArrays.split_plink(datadir * &quot;hapmap3&quot;, :chromosome; prefix = &quot;hapmap3.chr.&quot;)
# compresse Plink files for chromosome 1-23
for chr in 1:23
    plinkfile = &quot;hapmap3.chr.&quot; * string(chr)
    SnpArrays.compress_plink(plinkfile)
end
# delete uncompressed chromosome Plink files
for chr in 1:26
    plinkfile = &quot;hapmap3.chr.&quot; * string(chr)
    rm(plinkfile * &quot;.bed&quot;, force=true)
    rm(plinkfile * &quot;.bim&quot;, force=true)
    rm(plinkfile * &quot;.fam&quot;, force=true)
end
# copy covariate.txt file
cp(datadir * &quot;covariate.txt&quot;, joinpath(pwd(), &quot;covariate.txt&quot;))</code></pre><ul><li>The second script <a href="https://raw.githubusercontent.com/OpenMendel/OrdinalGWAS.jl/master/docs/cluster_run.jl"><code>cluster_run.jl</code></a> first fits the null model then submits a separate job for each chromosome. Run</li></ul><pre><code class="language-julia hljs">julia cluster_run.jl</code></pre><p>on head node.</p><pre><code class="language-julia hljs">run(`cat $datadir/../docs/cluster_run.jl`);</code></pre><pre><code class="nohighlight hljs">#!/usr/local/bin/julia
#
# This script demonstrates how to submit multiple OrdinalGWAS runs from multiple sets of
# Plink files on UCLA Hoffman2 cluster. It assumes that a demo data is available by
# running `julia cluster_preparedata.jl` at current folder.
#

using OrdinalGWAS, Serialization

# Step 1: fit null model and save result to file `fittednullmodel.jls`
nm = ordinalgwas(@formula(trait ~ sex), &quot;covariate.txt&quot;, nothing)
open(&quot;fittednullmodel.jls&quot;, &quot;w&quot;) do io
    Serialization.serialize(io, nm)
end

# Step 2: GWAS for each chromosome
for chr in 1:23
    println(&quot;submit job for chromosome=$chr&quot;)
    jcode = &quot;using OrdinalGWAS, Serialization;
    nm = open(deserialize, \&quot;fittednullmodel.jls\&quot;);
    bedfile = \&quot;hapmap3.chr.\&quot; * string($chr) * \&quot;.bed.gz\&quot;;
    bimfile = \&quot;hapmap3.chr.\&quot; * string($chr) * \&quot;.bim.gz\&quot;;
    pvalfile = \&quot;hapmap3.chr.\&quot; * string($chr) * \&quot;.pval.txt\&quot;;
    ordinalgwas(nm, bedfile, bimfile, 324; pvalfile=pvalfile);&quot;
    # prepare sh file for qsub
    open(&quot;tmp.sh&quot;, &quot;w&quot;) do io
        println(io, &quot;#!/bin/bash&quot;)
        println(io, &quot;#\$ -cwd&quot;)
        println(io, &quot;# error = Merged with joblog&quot;)
        println(io, &quot;#\$ -o joblog.\$JOB_ID&quot;)
        println(io, &quot;#\$ -j y&quot;)
        println(io, &quot;#\$ -l h_rt=0:30:00,h_data=2G&quot;) # request runtime and memory
        println(io, &quot;#\$ -pe shared 2&quot;) # request # shared-memory nodes
        println(io, &quot;# Email address to notify&quot;)
        println(io, &quot;#\$ -M \$USER@mail&quot;)
        println(io, &quot;# Notify when&quot;)
        println(io, &quot;#\$ -m a&quot;)
        println(io)
        println(io, &quot;# load the job environment:&quot;)
        println(io, &quot;. /u/local/Modules/default/init/modules.sh&quot;)
        println(io, &quot;module load julia/1.0.1&quot;) # available Julia version
        println(io)
        println(io, &quot;# run julia code&quot;)
        println(io, &quot;julia -e &#39;$jcode&#39; &gt; output.\$JOB_ID 2&gt;&amp;1&quot;)
    end
    # submit job
    run(`qsub tmp.sh`)
end</code></pre><h2 id="Proportional-Odds-Assumption"><a class="docs-heading-anchor" href="#Proportional-Odds-Assumption">Proportional Odds Assumption</a><a id="Proportional-Odds-Assumption-1"></a><a class="docs-heading-anchor-permalink" href="#Proportional-Odds-Assumption" title="Permalink"></a></h2><p>The ordered multinomial model used in this package is a cumulative logit model where one effect size is estimated for each covariate. Using the logit link yields the commonly used proportional odds model which assumes the proportional odds assumption holds for your data/model. The proportional odds assumption assumes that the effect of a covariate <span>$\boldsymbol{\beta}$</span> is constant across different groupings of the ordinal outcome i.e.</p><p>The SNP&#39;s effect (on the odds ratio) is the same for the odds ratio of mild vs {medium, severe} as it is for the odds ratio of {mild, medium} vs severe. </p><p>The consequences/neccesity of this assumption has been thoroughly discussed, as an example see <a href="https://www.fharrell.com/post/po/">this blog post</a>. In simualtions, we did not find violation of this to increase type I error. As stated in the post, when the assumption is violated (meaning there are different effect sizes for different groupings), the estimated effect size is like a weighted average of the different effect sizes. Formal tests for proportional odds are likely to reject the assumption for large n (see p. 306, Categorical Data Analysis 3rd edition by Agresti). Violation of this assumption means the effect size <span>$\boldsymbol{\beta}$</span> differs across outcome strata, but should not imply a significant finding is not significant.</p></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.15 on <span class="colophon-date" title="Thursday 31 March 2022 02:34">Thursday 31 March 2022</span>. Using Julia version 1.6.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
