<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>OrdinalGWAS.jl · OrdinalGWAS</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">OrdinalGWAS</span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>OrdinalGWAS.jl</a><ul class="internal"><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Example-data-set"><span>Example data set</span></a></li><li><a class="tocitem" href="#Basic-usage"><span>Basic usage</span></a></li><li><a class="tocitem" href="#Link-functions"><span>Link functions</span></a></li><li><a class="tocitem" href="#SNP-models"><span>SNP models</span></a></li><li><a class="tocitem" href="#SNP-and/or-sample-masks"><span>SNP and/or sample masks</span></a></li><li><a class="tocitem" href="#Likelihood-ratio-test-(LRT)"><span>Likelihood ratio test (LRT)</span></a></li><li><a class="tocitem" href="#Score-test-for-screening,-LRT-for-power"><span>Score test for screening, LRT for power</span></a></li><li><a class="tocitem" href="#GxE-or-other-interactions"><span>GxE or other interactions</span></a></li><li><a class="tocitem" href="#SNP-set-testing"><span>SNP-set testing</span></a></li><li><a class="tocitem" href="#Plotting-Results"><span>Plotting Results</span></a></li><li><a class="tocitem" href="#Docker"><span>Docker</span></a></li><li><a class="tocitem" href="#Multiple-Plink-file-sets"><span>Multiple Plink file sets</span></a></li><li><a class="tocitem" href="#Multiple-Plink-file-sets-on-cluster"><span>Multiple Plink file sets on cluster</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>OrdinalGWAS.jl</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>OrdinalGWAS.jl</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/OrdinalGWAS.jl/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="OrdinalGWAS.jl"><a class="docs-heading-anchor" href="#OrdinalGWAS.jl">OrdinalGWAS.jl</a><a id="OrdinalGWAS.jl-1"></a><a class="docs-heading-anchor-permalink" href="#OrdinalGWAS.jl" title="Permalink"></a></h1><p>OrdinalGWAS.jl is a Julia package for performing genome-wide association studies (GWAS) for ordered categorical phenotypes using <a href="https://en.wikipedia.org/wiki/Ordered_logit">proportional odds model</a> or <a href="https://en.wikipedia.org/wiki/Ordered_probit">ordred Probit model</a>. It is useful when the phenotype takes ordered discrete values, e.g., disease status (undiagnosed, pre-disease, mild, moderate, severe).</p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>This package requires Julia v0.7 or later and two other unregistered packages SnpArrays.jl and OrdinalMultinomialModels.jl. The package has not yet been registered and must be installed using the repository location. Start julia and use the ] key to switch to the package manager REPL</p><pre><code class="language-julia">(v1.1) pkg&gt; add https://github.com/OpenMendel/SnpArrays.jl
(v1.1) pkg&gt; add https://github.com/OpenMendel/OrdinalMultinomialModels.jl
(v1.1) pkg&gt; add https://github.com/OpenMendel/OrdinalGWAS.jl</code></pre><pre><code class="language-julia"># machine information for this tutorial
versioninfo()</code></pre><pre><code class="language-none">Julia Version 1.2.0
Commit c6da87ff4b (2019-08-20 00:03 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.6.0)
  CPU: Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, haswell)</code></pre><pre><code class="language-julia"># for use in this tutorial
using BenchmarkTools, CSV, Glob, OrdinalGWAS, SnpArrays</code></pre><h2 id="Example-data-set"><a class="docs-heading-anchor" href="#Example-data-set">Example data set</a><a id="Example-data-set-1"></a><a class="docs-heading-anchor-permalink" href="#Example-data-set" title="Permalink"></a></h2><p><code>data</code> folder of the package contains an example data set. In general, user can locate this folder by command</p><pre><code class="language-julia">using OrdinalGWAS
const datadir = normpath(joinpath(dirname(pathof(OrdinalGWAS)), &quot;../data/&quot;))</code></pre><pre><code class="language-none">&quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/&quot;</code></pre><pre><code class="language-julia"># content of the data folder
readdir(glob&quot;*.*&quot;, datadir)</code></pre><pre><code class="language-none">7-element Array{String,1}:
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/covariate.txt&quot;        
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.bed&quot;          
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.bim&quot;          
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.fam&quot;          
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.map&quot;          
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap_snpsetfile.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/simtrait.jl&quot;</code></pre><h2 id="Basic-usage"><a class="docs-heading-anchor" href="#Basic-usage">Basic usage</a><a id="Basic-usage-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-usage" title="Permalink"></a></h2><p>The following command performs GWAS using the <a href="https://en.wikipedia.org/wiki/Ordered_logit">proportional odds model</a>. The output is the fitted null model.</p><pre><code class="language-julia">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;)</code></pre><pre><code class="language-none">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><p>For documentation of the <code>ordinalgwas</code> function, type <code>?ordinalgwas</code> in Julia REPL.</p><article class="docstring"><header><a class="docstring-binding" id="OrdinalGWAS.ordinalgwas" href="#OrdinalGWAS.ordinalgwas"><code>OrdinalGWAS.ordinalgwas</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">ordinalgwas(nullformula, covfile, plinkfile)
ordinalgwas(nullformula, df, plinkfile)
ordinalgwas(fittednullmodel, plinkfile)
ordinalgwas(fittednullmodel, bedfile, bimfile, bedn)</code></pre><p><strong>Positional arguments</strong></p><ul><li><code>nullformula::FormulaTerm</code>: formula for the null model.</li><li><code>covfile::AbstractString</code>: covariate file (csv) with one header line. One column    should be the ordinal phenotype coded as integers starting from 1.  For example,    ordinal phenotypes can be coded as 1, 2, 3, 4 but not 0, 1, 2, 3.  </li><li><code>df::DataFrame</code>: DataFrame containing response and regressors for null model.</li><li><code>plinkfile::AbstractString</code>: Plink file name without the bed, fam, or bim    extensions. If <code>plinkfile==nothing</code>, only null model is fitted. If <code>plinkfile</code>    is provided, bed, bim, and fam file with same <code>plinkfile</code> prefix need to exist.    Compressed file formats such as gz and bz2 are allowed. Check all allowed formats    by <code>SnpArrays.ALLOWED_FORMAT</code>.  </li><li><code>fittednullmodel::StatsModels.TableRegressionModel</code>: the fitted null model    output from <code>ordinalgwas(nullformula, covfile)</code> or <code>ordinalgwas(nullformula, df)</code>.</li><li><code>bedfile::Union{AbstractString,IOStream}</code>: path to Plink bed file with full file name.</li><li><code>bimfile::Union{AbstractString,IOStream}</code>: path to Plink bim file with full file name.</li><li><code>bedn::Integer</code>: number of samples in bed file.</li></ul><p><strong>Keyword arguments</strong></p><ul><li><code>nullfile::Union{AbstractString, IOStream}</code>: output file for the fitted null model;    default is <code>ordinalgwas.null.txt</code>. </li><li><code>pvalfile::Union{AbstractString, IOStream}</code>: output file for the gwas p-values; default is    <code>ordinalgwas.pval.txt</code>. </li><li><code>covtype::Vector{DataType}</code>: type information for <code>covfile</code>. This is useful   when <code>CSV.read(covarfile)</code> has parsing errors.  </li><li><code>covrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for covariate file.  </li><li><code>testformula::FormulaTerm</code>: formula for test unit. Default is <code>@formula(trait ~ 0 + snp)</code>.</li><li><code>test::Symbol</code>: <code>:score</code> (default) or <code>:lrt</code>.  </li><li><code>link::GLM.Link</code>: <code>LogitLink()</code> (default), <code>ProbitLink()</code>, <code>CauchitLink()</code>,   or <code>CloglogLink()</code>.</li><li><code>snpmodel</code>: <code>ADDITIVE_MODEL</code> (default), <code>DOMINANT_MODEL</code>, or <code>RECESSIVE_MODEL</code>.</li><li><code>snpinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: SNP indices for bed file.</li><li><code>bedrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for bed file.</li><li><code>solver</code>: an optimization solver supported by MathProgBase. Default is    <code>NLoptSolver(algorithm=:LD_SLSQP, maxeval=4000)</code>. Another common choice is    <code>IpoptSolver(print_level=0)</code>.</li><li><code>verbose::Bool</code>: default is <code>false</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/OrdinalGWAS.jl/blob/1ecbc8ac64e47163129200ea7c1163ee9f12bf10/src/gwas.jl#L1-L43">source</a></section></article><h3 id="Formula-for-null-model"><a class="docs-heading-anchor" href="#Formula-for-null-model">Formula for null model</a><a id="Formula-for-null-model-1"></a><a class="docs-heading-anchor-permalink" href="#Formula-for-null-model" title="Permalink"></a></h3><p>The first argument specifies the null model without SNP effects, e.g., <code>@formula(trait ~ sex)</code>.</p><h3 id="Input-files"><a class="docs-heading-anchor" href="#Input-files">Input files</a><a id="Input-files-1"></a><a class="docs-heading-anchor-permalink" href="#Input-files" title="Permalink"></a></h3><p><code>ordinalgwas</code> expects two input files: one for responses plus covariates (second argument), the other the Plink files for genotypes (third argument).</p><h4 id="Covariate-and-trait-file"><a class="docs-heading-anchor" href="#Covariate-and-trait-file">Covariate and trait file</a><a id="Covariate-and-trait-file-1"></a><a class="docs-heading-anchor-permalink" href="#Covariate-and-trait-file" title="Permalink"></a></h4><p>Covariates and phenotype are provided in a csv file, e.g., <code>covariate.txt</code>, which has one header line for variable names. In this example, variable <code>trait</code> is the ordered categorical phenotypes coded as integers 1 to 4. We want to include variable <code>sex</code> as the covariate in GWAS.</p><pre><code class="language-julia">run(`head $(datadir)covariate.txt`);</code></pre><pre><code class="language-none">famid,perid,faid,moid,sex,trait
2431,NA19916,0,0,1,4
2424,NA19835,0,0,2,4
2469,NA20282,0,0,2,4
2368,NA19703,0,0,1,3
2425,NA19901,0,0,2,3
2427,NA19908,0,0,1,4
2430,NA19914,0,0,2,4
2470,NA20287,0,0,2,1
2436,NA19713,0,0,2,3</code></pre><h4 id="Plink-file"><a class="docs-heading-anchor" href="#Plink-file">Plink file</a><a id="Plink-file-1"></a><a class="docs-heading-anchor-permalink" href="#Plink-file" title="Permalink"></a></h4><p>Genotype data is available as binary Plink files.</p><pre><code class="language-julia">readdir(glob&quot;hapmap3.*&quot;, datadir)</code></pre><pre><code class="language-none">4-element Array{String,1}:
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.bed&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.bim&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.fam&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.map&quot;</code></pre><p>In this example, there are 324 samples at 13,928 SNPs.</p><pre><code class="language-julia">size(SnpArray(datadir * &quot;hapmap3.bed&quot;))</code></pre><pre><code class="language-none">(324, 13928)</code></pre><p>Compressed Plink files are supported. For example, if Plink files are <code>hapmap3.bed.gz</code>, <code>hapmap3.bim.gz</code> and <code>hapmap3.fam.gz</code>, the same command</p><pre><code class="language-julia">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;)</code></pre><p>still works. Check all supported compression format by</p><pre><code class="language-julia">SnpArrays.ALLOWED_FORMAT</code></pre><pre><code class="language-none">6-element Array{String,1}:
 &quot;gz&quot;  
 &quot;zlib&quot;
 &quot;zz&quot;  
 &quot;xz&quot;  
 &quot;zst&quot; 
 &quot;bz2&quot;</code></pre><h3 id="Output-files"><a class="docs-heading-anchor" href="#Output-files">Output files</a><a id="Output-files-1"></a><a class="docs-heading-anchor-permalink" href="#Output-files" title="Permalink"></a></h3><p><code>ordinalgwas</code> outputs two files: <code>ordinalgwas.null.txt</code> and <code>ordinalgwas.pval.txt</code>. </p><ul><li><code>ordinalgwas.null.txt</code> lists the estimated null model (without SNPs). </li></ul><pre><code class="language-julia">run(`cat ordinalgwas.null.txt`);</code></pre><pre><code class="language-none">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><ul><li><code>ordinalgwas.pval.txt</code> tallies the SNPs and their pvalues. </li></ul><pre><code class="language-julia">run(`head ordinalgwas.pval.txt`);</code></pre><pre><code class="language-none">chr,pos,snpid,maf,hwepval,pval
1,554484,rs10458597,0.0,1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.40987633326666817,0.004565312839540994
1,967643,rs2710875,0.32407407407407407,4.0762491007057454e-7,3.108283828554874e-5
1,1168108,rs11260566,0.19158878504672894,0.12856822794469086,1.2168672367668889e-5
1,1375074,rs1312568,0.441358024691358,2.537601965061486e-19,0.008206860046175225
1,1588771,rs35154105,0.0,1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,0.5111981332544
1,1990452,rs2678939,0.4537037037037037,5.0769595770843126e-11,0.2997282957184783
1,2194615,rs7553178,0.22685185185185186,0.17056143157457782,0.1713331245805063</code></pre><p>Output file names can be changed by the <code>nullfile</code> and <code>pvalfile</code> keywords respectively. For example, </p><pre><code class="language-julia">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, pvalfile=&quot;ordinalgwas.pval.txt.gz&quot;)</code></pre><p>will output the p-value file in compressed gz format.</p><h3 id="Subsamples"><a class="docs-heading-anchor" href="#Subsamples">Subsamples</a><a id="Subsamples-1"></a><a class="docs-heading-anchor-permalink" href="#Subsamples" title="Permalink"></a></h3><p>Use the keyword <code>covrowinds</code> to specify selected samples in the covarite file. Use the keyword <code>bedrowinds</code> to specify selected samples in the Plink bed file. For example, to use the first 300 samples in both covariate and bed file:</p><pre><code class="language-julia">ordinalgwas(@formula(trait ~ sex), covfile, plkfile, covrowinds=1:300, bedrowinds=1:300)</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Users should always make sure that the selected samples in covariate file match exactly those in bed file. </p></div></div><h3 id="Input-non-genetic-data-as-DataFrame"><a class="docs-heading-anchor" href="#Input-non-genetic-data-as-DataFrame">Input non-genetic data as DataFrame</a><a id="Input-non-genetic-data-as-DataFrame-1"></a><a class="docs-heading-anchor-permalink" href="#Input-non-genetic-data-as-DataFrame" title="Permalink"></a></h3><p>Internally <code>ordinalgwas</code> parses the covariate file as a DataFrame by <code>CSV.read(covfile)</code>. For covariate file of other formats, users can parse it as a DataFrame and then input the DataFrame to <code>ordinalgwas</code> directly.</p><pre><code class="language-julia">ordinalgwas(@formula(trait ~ sex), df, plinkfile)</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Users should always make sure that individuals in covariate file or DataFrame match those in Plink fam file. </p></div></div><p>For example, following code checks that the first 2 columns of the <code>covariate.txt</code> file match the first 2 columns of the <code>hapmap3.fam</code> file exactly.</p><pre><code class="language-julia">covdf = CSV.read(datadir * &quot;covariate.txt&quot;)
plkfam = CSV.read(datadir * &quot;hapmap3.fam&quot;, header=0, delim=&#39; &#39;)
all(covdf[!, 1] .== plkfam[!, 1]) &amp;&amp; all(covdf[!, 2] .== plkfam[!, 2])</code></pre><pre><code class="language-none">true</code></pre><h3 id="Timing"><a class="docs-heading-anchor" href="#Timing">Timing</a><a id="Timing-1"></a><a class="docs-heading-anchor-permalink" href="#Timing" title="Permalink"></a></h3><p>For this moderate-sized data set, <code>ordinalgwas</code> takes around 0.2 seconds.</p><pre><code class="language-julia">@btime(ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;));</code></pre><pre><code class="language-none">  210.824 ms (639565 allocations: 32.70 MiB)</code></pre><pre><code class="language-julia"># clean up
rm(&quot;ordinalgwas.null.txt&quot;, force=true)
rm(&quot;ordinalgwas.pval.txt&quot;, force=true)</code></pre><h2 id="Link-functions"><a class="docs-heading-anchor" href="#Link-functions">Link functions</a><a id="Link-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Link-functions" title="Permalink"></a></h2><p>The <code>link</code> keyword argument of <code>ordinalgwas</code> can take value:  </p><ul><li><code>LogitLink()</code>, proportional odds model (default),  </li><li><code>ProbitLink()</code>, ordred Probit model,  </li><li><code>CloglogLink()</code>, proportional hazards model, or </li><li><code>CauchyLink()</code>.</li></ul><p>For example, to perform GWAS using the ordred Probit model</p><pre><code class="language-julia">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    link=ProbitLink(), nullfile=&quot;opm.null.txt&quot;, pvalfile=&quot;opm.pval.txt&quot;)</code></pre><pre><code class="language-none">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,ProbitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -0.866156   0.210677  -4.11129    &lt;1e-4 
intercept2|3  -0.359878   0.205817  -1.74854    0.0813
intercept3|4   0.247054   0.205382   1.2029     0.2299
sex            0.251058   0.128225   1.95795    0.0511
──────────────────────────────────────────────────────</code></pre><p>The estimates in null model and p-values are slightly different from those in proportional odds moodel.</p><pre><code class="language-julia">run(`head opm.pval.txt`);</code></pre><pre><code class="language-none">chr,pos,snpid,maf,hwepval,pval
1,554484,rs10458597,0.0,1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.40987633326666817,0.010076916742300138
1,967643,rs2710875,0.32407407407407407,4.0762491007057454e-7,2.6272564941853933e-5
1,1168108,rs11260566,0.19158878504672894,0.12856822794469086,1.089748485107844e-5
1,1375074,rs1312568,0.441358024691358,2.537601965061486e-19,0.005102883990438149
1,1588771,rs35154105,0.0,1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,0.48653776297859236
1,1990452,rs2678939,0.4537037037037037,5.0769595770843126e-11,0.33231290090455434
1,2194615,rs7553178,0.22685185185185186,0.17056143157457782,0.2591551397719743</code></pre><pre><code class="language-julia">rm(&quot;opm.null.txt&quot;, force=true)
rm(&quot;opm.pval.txt&quot;, force=true)</code></pre><h2 id="SNP-models"><a class="docs-heading-anchor" href="#SNP-models">SNP models</a><a id="SNP-models-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-models" title="Permalink"></a></h2><p>Genotypes are translated into numeric values according to different genetic model, which is specified by the <code>snpmodel</code> keyword. Default is <code>ADDITIVE_MODEL</code>.</p><table><tr><th style="text-align: center">Genotype</th><th style="text-align: center"><code>SnpArray</code></th><th style="text-align: center"><code>ADDITIVE_MODEL</code></th><th style="text-align: center"><code>DOMINANT_MODEL</code></th><th style="text-align: center"><code>RECESSIVE_MODEL</code></th></tr><tr><td style="text-align: center">A1,A1</td><td style="text-align: center">0x00</td><td style="text-align: center">0</td><td style="text-align: center">0</td><td style="text-align: center">0</td></tr><tr><td style="text-align: center">missing</td><td style="text-align: center">0x01</td><td style="text-align: center">NaN</td><td style="text-align: center">NaN</td><td style="text-align: center">NaN</td></tr><tr><td style="text-align: center">A1,A2</td><td style="text-align: center">0x02</td><td style="text-align: center">1</td><td style="text-align: center">1</td><td style="text-align: center">0</td></tr><tr><td style="text-align: center">A2,A2</td><td style="text-align: center">0x03</td><td style="text-align: center">2</td><td style="text-align: center">1</td><td style="text-align: center">1</td></tr></table><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>ordinalgwas</code> imputes missing genotypes according to minor allele frequencies. </p></div></div><p>Users are advised to impute genotypes using more sophiscated methods before GWAS.</p><h2 id="SNP-and/or-sample-masks"><a class="docs-heading-anchor" href="#SNP-and/or-sample-masks">SNP and/or sample masks</a><a id="SNP-and/or-sample-masks-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-and/or-sample-masks" title="Permalink"></a></h2><p>In practice, we often perform GWAS on selected SNPs and/or selected samples. They can be specified by the <code>snpinds</code>, <code>covrowinds</code> and <code>bedrowinds</code> keywords of <code>ordinalgwas</code> function. </p><p>For example, to perform GWAS on SNPs with minor allele frequency (MAF) above 0.05</p><pre><code class="language-julia"># create SNP mask
snpinds = maf(SnpArray(&quot;../data/hapmap3.bed&quot;)) .≥ 0.05
# GWAS on selected SNPs
@time ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    snpinds=snpinds, nullfile=&quot;commonvariant.null.txt&quot;, pvalfile=&quot;commonvariant.pval.txt&quot;)</code></pre><pre><code class="language-none">  0.384494 seconds (946.69 k allocations: 49.160 MiB, 3.73% gc time)





StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia">run(`head commonvariant.pval.txt`);</code></pre><pre><code class="language-none">chr,pos,snpid,maf,hwepval,pval
1,758311,rs12562034,0.07763975155279501,0.40987633326666817,0.004565312839540994
1,967643,rs2710875,0.32407407407407407,4.0762491007057454e-7,3.108283828554874e-5
1,1168108,rs11260566,0.19158878504672894,0.12856822794469086,1.2168672367668889e-5
1,1375074,rs1312568,0.441358024691358,2.537601965061486e-19,0.008206860046175225
1,1990452,rs2678939,0.4537037037037037,5.0769595770843126e-11,0.2997282957184783
1,2194615,rs7553178,0.22685185185185186,0.17056143157457782,0.1713331245805063
1,2396747,rs13376356,0.1448598130841121,0.905307921507814,0.5320416198875456
1,2823603,rs1563468,0.4830246913580247,4.2306553724392545e-9,0.225191391783573
1,3025087,rs6690373,0.2538699690402477,9.23864188719278e-8,0.7018469417717486</code></pre><pre><code class="language-julia"># extra header line in commonvariant.pval.txt
countlines(&quot;commonvariant.pval.txt&quot;), count(snpinds)</code></pre><pre><code class="language-none">(12086, 12085)</code></pre><pre><code class="language-julia"># clean up
rm(&quot;commonvariant.null.txt&quot;, force=true)
rm(&quot;commonvariant.pval.txt&quot;, force=true)</code></pre><p><code>covrowinds</code> specify the samples in the covariate file and <code>bedrowinds</code> for SnpArray. User should be particularly careful when using these two keyword. Selected rows in SnpArray should exactly match the samples in the null model. Otherwise the results are meaningless.</p><h2 id="Likelihood-ratio-test-(LRT)"><a class="docs-heading-anchor" href="#Likelihood-ratio-test-(LRT)">Likelihood ratio test (LRT)</a><a id="Likelihood-ratio-test-(LRT)-1"></a><a class="docs-heading-anchor-permalink" href="#Likelihood-ratio-test-(LRT)" title="Permalink"></a></h2><p>By default, <code>ordinalgwas</code> calculates p-value for each SNP using score test. Score test is fast because it doesn&#39;t require fitting alternative model for each SNP. User can request likelihood ratio test (LRT) using keyword <code>test=:lrt</code>. LRT is much slower but may be more powerful than score test.</p><pre><code class="language-julia">@time ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    test=:LRT, nullfile=&quot;lrt.null.txt&quot;, pvalfile=&quot;lrt.pval.txt&quot;)</code></pre><pre><code class="language-none"> 27.756967 seconds (8.59 M allocations: 2.106 GiB, 1.77% gc time)





StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><p>Note the extra <code>effect</code> column in pvalfile, which is the effect size (regression coefficient) for each SNP. </p><pre><code class="language-julia">run(`head lrt.pval.txt`);</code></pre><pre><code class="language-none">chr,pos,snpid,maf,hwepval,effect,pval
1,554484,rs10458597,0.0,1.0,0.0,1.0
1,758311,rs12562034,0.07763975155279501,0.40987633326666817,-1.0057833719544331,0.0019185836579804134
1,967643,rs2710875,0.32407407407407407,4.0762491007057454e-7,-0.6488560566295055,1.805050556976241e-5
1,1168108,rs11260566,0.19158878504672894,0.12856822794469086,-0.9157225669357879,5.873384712685568e-6
1,1375074,rs1312568,0.441358024691358,2.537601965061486e-19,-0.33181366525772593,0.008081022577832324
1,1588771,rs35154105,0.0,1.0,0.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,-0.7338026388701573,0.5169027130129711
1,1990452,rs2678939,0.4537037037037037,5.0769595770843126e-11,-0.13586499231819726,0.29946402200912603
1,2194615,rs7553178,0.22685185185185186,0.17056143157457782,-0.2512075640440123,0.16151069094439868</code></pre><pre><code class="language-julia"># clean up
rm(&quot;lrt.pval.txt&quot;, force=true)
rm(&quot;lrt.null.txt&quot;, force=true)</code></pre><p>In this example, GWAS by score test takes less than 0.2 second, while GWAS by LRT takes about 20 seconds. About 100 fold difference in run time. </p><h2 id="Score-test-for-screening,-LRT-for-power"><a class="docs-heading-anchor" href="#Score-test-for-screening,-LRT-for-power">Score test for screening, LRT for power</a><a id="Score-test-for-screening,-LRT-for-power-1"></a><a class="docs-heading-anchor-permalink" href="#Score-test-for-screening,-LRT-for-power" title="Permalink"></a></h2><p>For large data sets, a practical solution is to perform score test first, then re-do LRT for the most promising SNPs according to score test p-values.</p><p><strong>Step 1</strong>: Perform score test GWAS, results in <code>score.pval.txt</code>.</p><pre><code class="language-julia">@time ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    test=:score, pvalfile=&quot;score.pval.txt&quot;);</code></pre><pre><code class="language-none">  0.338187 seconds (668.69 k allocations: 34.240 MiB, 6.17% gc time)</code></pre><pre><code class="language-julia">run(`head score.pval.txt`);</code></pre><pre><code class="language-none">chr,pos,snpid,maf,hwepval,pval
1,554484,rs10458597,0.0,1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.40987633326666817,0.004565312839540994
1,967643,rs2710875,0.32407407407407407,4.0762491007057454e-7,3.108283828554874e-5
1,1168108,rs11260566,0.19158878504672894,0.12856822794469086,1.2168672367668889e-5
1,1375074,rs1312568,0.441358024691358,2.537601965061486e-19,0.008206860046175225
1,1588771,rs35154105,0.0,1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,0.5111981332544
1,1990452,rs2678939,0.4537037037037037,5.0769595770843126e-11,0.2997282957184783
1,2194615,rs7553178,0.22685185185185186,0.17056143157457782,0.1713331245805063</code></pre><p><strong>Step 2</strong>: Sort score test p-values and find top 10 SNPs.</p><pre><code class="language-julia">scorepvals = CSV.read(&quot;score.pval.txt&quot;)[!, 6] # p-values in 5th column
tophits = sortperm(scorepvals)[1:10] # indices of 10 SNPs with smallest p-values
scorepvals[tophits] # smallest 10 p-values</code></pre><pre><code class="language-none">10-element Array{Float64,1}:
 1.3080149099181303e-6
 6.536722765052079e-6 
 9.66474218566903e-6  
 1.2168672367668889e-5
 1.8027460018331254e-5
 2.0989542284213636e-5
 2.6844521269963608e-5
 3.108283828554874e-5 
 4.1010912875160476e-5
 4.2966265138454725e-5</code></pre><p><strong>Step 3</strong>: Re-do LRT on top hits.</p><pre><code class="language-julia">@time ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    snpinds=tophits, test=:LRT, pvalfile=&quot;lrt.pval.txt&quot;);</code></pre><pre><code class="language-none">  0.345620 seconds (512.98 k allocations: 27.686 MiB, 4.47% gc time)</code></pre><pre><code class="language-julia">run(`cat lrt.pval.txt`);</code></pre><pre><code class="language-none">chr,pos,snpid,maf,hwepval,effect,pval
1,967643,rs2710875,0.32407407407407407,4.0762491007057454e-7,-0.6488560566295055,1.805050556976241e-5
1,1168108,rs11260566,0.19158878504672894,0.12856822794469086,-0.9157225669357879,5.873384712685568e-6
3,36821790,rs4678553,0.23456790123456794,0.10946682163244979,0.7424952268973518,1.1303825016262592e-5
4,11017683,rs16881446,0.27554179566563464,0.8942746118760273,-0.7870581482955515,1.1105427468799613e-5
5,3739190,rs12521166,0.0679012345679012,0.18613647746093892,1.1468852997925316,4.781288229657399e-5
6,7574576,rs1885466,0.17746913580246915,0.762068717898719,0.8750621092263019,7.272346896740631e-6
6,52474721,rs2073183,0.1826625386996904,0.5077765730476698,0.7790794914858663,5.069394513906121e-5
7,41152376,rs28880,0.3379629629629629,0.8052368892744095,-0.814633902445351,9.180126530294943e-7
7,84223996,rs4128623,0.07870370370370372,0.021834717346756814,1.0022229316338573,6.587895464657512e-5
23,121048059,rs1937165,0.4380804953560371,3.959609737265111e-16,0.5392313636256612,1.9754643855522616e-5</code></pre><pre><code class="language-julia"># clean up
rm(&quot;ordinalgwas.null.txt&quot;, force=true)
rm(&quot;score.pval.txt&quot;, force=true)
rm(&quot;lrt.pval.txt&quot;, force=true)</code></pre><h2 id="GxE-or-other-interactions"><a class="docs-heading-anchor" href="#GxE-or-other-interactions">GxE or other interactions</a><a id="GxE-or-other-interactions-1"></a><a class="docs-heading-anchor-permalink" href="#GxE-or-other-interactions" title="Permalink"></a></h2><p>In many applications, we want to test SNP effect and/or its interaction with other terms. <code>testformula</code> keyword specifies the test unit <strong>besides</strong> the covariates in <code>nullformula</code>. </p><p>In following example, keyword <code>testformula=@formula(trait ~ snp + snp &amp; sex)</code> instructs <code>ordinalgwas</code> to test joint effect of <code>snp</code> and <code>snp &amp; sex</code> interaction.</p><pre><code class="language-julia">ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;, 
    pvalfile=&quot;GxE.pval.txt&quot;, testformula=@formula(trait ~ snp + snp &amp; sex));</code></pre><pre><code class="language-julia">run(`head GxE.pval.txt`);</code></pre><pre><code class="language-none">chr,pos,snpid,maf,hwepval,pval
1,554484,rs10458597,0.0,1.0,1.0
1,758311,rs12562034,0.07763975155279501,0.40987633326666817,0.01744601041225423
1,967643,rs2710875,0.32407407407407407,4.0762491007057454e-7,0.0001667073239489097
1,1168108,rs11260566,0.19158878504672894,0.12856822794469086,4.763762457893383e-5
1,1375074,rs1312568,0.441358024691358,2.537601965061486e-19,0.02913847124299368
1,1588771,rs35154105,0.0,1.0,1.0
1,1789051,rs16824508,0.00462962962962965,0.9332783156468178,0.29643631149443217
1,1990452,rs2678939,0.4537037037037037,5.0769595770843126e-11,0.37924580479348796
1,2194615,rs7553178,0.22685185185185186,0.17056143157457782,0.325582269932396</code></pre><pre><code class="language-julia"># clean up
rm(&quot;ordinalgwas.null.txt&quot;)
rm(&quot;GxE.pval.txt&quot;)</code></pre><p>For some applications, the user may want to simply test the GxE interaction effect. This requires fitting the SNP in the null model and is quite slower, but the command <code>ordinalgwasGxE()</code> can be used test the interaction effect. To do this you must specify the environmental variable in the command, either as a symbol, such as &quot;:age&quot; or as a string &quot;age&quot;. </p><p>For documentation of the <code>ordinalgwasGxE</code> function, type <code>?ordinalgwasGxE</code> in Julia REPL.</p><article class="docstring"><header><a class="docstring-binding" id="OrdinalGWAS.ordinalgwasGxE" href="#OrdinalGWAS.ordinalgwasGxE"><code>OrdinalGWAS.ordinalgwasGxE</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">ordinalgwasGxE(nullformula, covfile, plinkfile, e)
ordinalgwasGxE(nullformula, df, plinkfile, e)
ordinalgwasGxE(fittednullmodel, plinkfile, e)
ordinalgwasGxE(fittednullmodel, bedfile, bimfile, bedn, e)</code></pre><p><strong>Positional arguments</strong></p><ul><li><code>nullformula::FormulaTerm</code>: formula for the null model.</li><li><code>covfile::AbstractString</code>: covariate file (csv) with one header line. One column    should be the ordinal phenotype coded as integers starting from 1.  For example,    ordinal phenotypes can be coded as 1, 2, 3, 4 but not 0, 1, 2, 3.  </li><li><code>df::DataFrame</code>: DataFrame containing response and regressors for null model.</li><li><code>plinkfile::AbstractString</code>: Plink file name without the bed, fam, or bim    extensions. If <code>plinkfile==nothing</code>, only null model is fitted. If <code>plinkfile</code>    is provided, bed, bim, and fam file with same <code>plinkfile</code> prefix need to exist.    Compressed file formats such as gz and bz2 are allowed. Check all allowed formats    by <code>SnpArrays.ALLOWED_FORMAT</code>.  </li><li><code>e::Union{AbstractString,Symbol}</code>: Enviromental variable to be used to test the GxE interaction.</li></ul><p>For instance, for testing <code>sex &amp; snp</code> interaction, use <code>:sex</code> or <code>&quot;sex&quot;</code>.</p><ul><li><code>fittednullmodel::StatsModels.TableRegressionModel</code>: the fitted null model    output from <code>ordinalgwas(nullformula, covfile)</code> or <code>ordinalgwas(nullformula, df)</code>.</li><li><code>bedfile::Union{AbstractString,IOStream}</code>: path to Plink bed file with full file name.</li><li><code>bimfile::Union{AbstractString,IOStream}</code>: path to Plink bim file with full file name.</li><li><code>bedn::Integer</code>: number of samples in bed file.</li></ul><p><strong>Keyword arguments</strong></p><ul><li><code>nullfile::Union{AbstractString, IOStream}</code>: output file for the fitted null model;    default is <code>ordinalgwas.null.txt</code>. </li><li><code>pvalfile::Union{AbstractString, IOStream}</code>: output file for the gwas p-values; default is    <code>ordinalgwas.pval.txt</code>. </li><li><code>covtype::Vector{DataType}</code>: type information for <code>covfile</code>. This is useful   when <code>CSV.read(covarfile)</code> has parsing errors.  </li><li><code>covrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for covariate file.  </li><li><code>test::Symbol</code>: <code>:score</code> (default) or <code>:lrt</code>.  </li><li><code>link::GLM.Link</code>: <code>LogitLink()</code> (default), <code>ProbitLink()</code>, <code>CauchitLink()</code>,   or <code>CloglogLink()</code>.</li><li><code>snpmodel</code>: <code>ADDITIVE_MODEL</code> (default), <code>DOMINANT_MODEL</code>, or <code>RECESSIVE_MODEL</code>.</li><li><code>snpinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: SNP indices for bed file.</li><li><code>bedrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for bed file.</li><li><code>solver</code>: an optimization solver supported by MathProgBase. Default is    <code>NLoptSolver(algorithm=:LD_SLSQP, maxeval=4000)</code>. Another common choice is    <code>IpoptSolver(print_level=0)</code>.</li><li><code>verbose::Bool</code>: default is <code>false</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/OrdinalGWAS.jl/blob/1ecbc8ac64e47163129200ea7c1163ee9f12bf10/src/gwas.jl#L598-L641">source</a></section></article><pre><code class="language-julia">ordinalgwasGxE(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;,
    :sex, pvalfile = &quot;gxe_snp.pval.txt&quot;, snpinds=1:5, test=:score)</code></pre><pre><code class="language-none">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia">run(`head gxe_snp.pval.txt`);</code></pre><pre><code class="language-none">chr,pos,snpid,maf,hwepval,snpeffectnull,pval
1,554484,rs10458597,0.0,1.0,0.0,1.0
1,758311,rs12562034,0.07763975155279501,0.40987633326666817,-1.0057833719544331,0.6377422425977491
1,967643,rs2710875,0.32407407407407407,4.0762491007057454e-7,-0.6488560566295055,0.9667114198051628
1,1168108,rs11260566,0.19158878504672894,0.12856822794469086,-0.9157225669357879,0.26352674694121003
1,1375074,rs1312568,0.441358024691358,2.537601965061486e-19,-0.33181366525772593,0.7811133315582837</code></pre><pre><code class="language-julia"># clean up
rm(&quot;gxe_snp.pval.txt&quot;, force=true)</code></pre><h2 id="SNP-set-testing"><a class="docs-heading-anchor" href="#SNP-set-testing">SNP-set testing</a><a id="SNP-set-testing-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-set-testing" title="Permalink"></a></h2><p>In many applications, we want to test a SNP-set. The function <code>ordinalsnpsetgwas()</code> can be used to do this. The snpset can be specified as either a window (test every X snps), a filename that specifies an input file, with no header and two columns separated by a space. The first column must contain the snpset ID and the second column must contain the snpid&#39;s identical to the bimfile, or an AbstractVector that allows you to specify the snps you want to perform one joint snpset test for.</p><p>For documentation of the <code>ordinalsnpsetgwas</code> function, type <code>?ordinalsnpsetgwas</code> in Julia REPL.</p><article class="docstring"><header><a class="docstring-binding" id="OrdinalGWAS.ordinalsnpsetgwas" href="#OrdinalGWAS.ordinalsnpsetgwas"><code>OrdinalGWAS.ordinalsnpsetgwas</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">ordinalsnpsetgwas(nullformula, covfile, plinkfile)
ordinalsnpsetgwas(nullformula, df, plinkfile)
ordinalsnpsetgwas(fittednullmodel, plinkfile)
ordinalsnpsetgwas(fittednullmodel, bedfile, bimfile, bedn)</code></pre><p><strong>Positional arguments</strong></p><ul><li><code>nullformula::FormulaTerm</code>: formula for the null model.</li><li><code>covfile::AbstractString</code>: covariate file (csv) with one header line. One column    should be the ordinal phenotype coded as integers starting from 1.  For example,    ordinal phenotypes can be coded as 1, 2, 3, 4 but not 0, 1, 2, 3.  </li><li><code>df::DataFrame</code>: DataFrame containing response and regressors for null model.</li><li><code>plinkfile::AbstractString</code>: Plink file name without the bed, fam, or bim    extensions. If <code>plinkfile==nothing</code>, only null model is fitted. If <code>plinkfile</code>    is provided, bed, bim, and fam file with same <code>plinkfile</code> prefix need to exist.    Compressed file formats such as gz and bz2 are allowed. Check all allowed formats    by <code>SnpArrays.ALLOWED_FORMAT</code>.  </li><li><code>fittednullmodel::StatsModels.TableRegressionModel</code>: the fitted null model    output from <code>ordinalgwas(nullformula, covfile)</code> or <code>ordinalgwas(nullformula, df)</code>.</li><li><code>bedfile::Union{AbstractString,IOStream}</code>: path to Plink bed file with full file name.</li><li><code>bimfile::Union{AbstractString,IOStream}</code>: path to Plink bim file with full file name.</li><li><code>bedn::Integer</code>: number of samples in bed file.</li></ul><p><strong>Keyword arguments</strong></p><ul><li><code>nullfile::Union{AbstractString, IOStream}</code>: output file for the fitted null model;    default is <code>ordinalgwas.null.txt</code>. </li><li><code>pvalfile::Union{AbstractString, IOStream}</code>: output file for the gwas p-values; default is    <code>ordinalgwas.pval.txt</code>. </li><li><code>snpset::Union{Nothing, Integer, AbstractString, AbstractVector{&lt;:Integer}}</code>: Only include    if you are conducting a snpset analysis. An integer indicates a window of SNPs    (i.e. every 500 snps). An abstract string allows you to specify an input file,    with no header and two columns separated by a space. The first column must contain the snpset ID   and the second column must contain the snpid&#39;s identical to the bimfile. An AbstractVector   allows you to specify the snps you want to perform one joint snpset test for.</li><li><code>covtype::Vector{DataType}</code>: type information for <code>covfile</code>. This is useful   when <code>CSV.read(covarfile)</code> has parsing errors.  </li><li><code>covrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for covariate file.  </li><li><code>test::Symbol</code>: <code>:score</code> (default) or <code>:lrt</code>.  </li><li><code>link::GLM.Link</code>: <code>LogitLink()</code> (default), <code>ProbitLink()</code>, <code>CauchitLink()</code>,   or <code>CloglogLink()</code>.</li><li><code>snpmodel</code>: <code>ADDITIVE_MODEL</code> (default), <code>DOMINANT_MODEL</code>, or <code>RECESSIVE_MODEL</code>.</li><li><code>snpinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: SNP indices for bed file.</li><li><code>bedrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for bed file.</li><li><code>solver</code>: an optimization solver supported by MathProgBase. Default is    <code>NLoptSolver(algorithm=:LD_SLSQP, maxeval=4000)</code>. Another common choice is    <code>IpoptSolver(print_level=0)</code>.</li><li><code>verbose::Bool</code>: default is <code>false</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/OrdinalGWAS.jl/blob/1ecbc8ac64e47163129200ea7c1163ee9f12bf10/src/gwas.jl#L262-L309">source</a></section></article><p>In following example, we perform a SNP-set test on the 50th to 55th snps. </p><pre><code class="language-julia">ordinalsnpsetgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;,
    pvalfile = &quot;snpset.pval.txt&quot;, snpset = 50:55)</code></pre><pre><code class="language-none">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia">run(`head snpset.pval.txt`);</code></pre><pre><code class="language-none">The joint pvalue of snps indexed at 50:55 is 0.36471265366639605</code></pre><pre><code class="language-julia"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;ordinalgwas.null.txt&quot;, force=true)</code></pre><p>In the following example we run a SNP-set test on the annotated SNP-set file.</p><pre><code class="language-julia">run(`head ../data/hapmap_snpsetfile.txt`);</code></pre><pre><code class="language-none">gene1 rs10458597
gene1 rs12562034
gene1 rs2710875
gene1 rs11260566
gene1 rs1312568
gene1 rs35154105
gene1 rs16824508
gene1 rs2678939
gene1 rs7553178
gene1 rs13376356</code></pre><pre><code class="language-julia">ordinalsnpsetgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;,
    pvalfile = &quot;snpset.pval.txt&quot;, snpset = datadir * &quot;/hapmap_snpsetfile.txt&quot;)</code></pre><pre><code class="language-none">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia">run(`head snpset.pval.txt`);</code></pre><pre><code class="language-none">snpsetid,nsnps,pval
gene1,93,1.72133946991466e-5
gene2,93,0.03692497684330217
gene3,93,0.7478549371392197
gene4,92,0.027650798223816195
gene5,93,0.6119581594582347
gene6,93,0.029184642230868463
gene7,93,0.3916007348891039
gene8,93,0.1253957410986018
gene9,93,0.7085635621609013</code></pre><pre><code class="language-julia"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;ordinalgwas.null.txt&quot;, force=true)</code></pre><p>In the following example we run a SNP-set test on every 15 SNPs.</p><pre><code class="language-julia">ordinalsnpsetgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, datadir * &quot;hapmap3&quot;,
    pvalfile = &quot;snpset.pval.txt&quot;, snpset=15)</code></pre><pre><code class="language-none">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><pre><code class="language-julia">run(`head snpset.pval.txt`);</code></pre><pre><code class="language-none">startchr,startpos,startsnpid,endchr,endpos,endsnpid,pval
1,554484,rs10458597,1,3431124,rs12093117,1.9394189465782687e-13
1,3633945,rs10910017,1,6514524,rs932112,0.11077869162820388
1,6715827,rs441515,1,9534606,rs4926480,0.2742450817971269
1,9737551,rs12047054,1,12559747,rs4845907,0.49346113650468615
1,12760427,rs848577,1,16021797,rs6679870,0.15447358658305502
1,16228774,rs1972359,1,19100349,rs9426794,0.1544232917025094
1,19301516,rs4912046,1,22122176,rs9426785,0.474987334946532
1,22323074,rs2235529,1,25166528,rs4648890,0.41246096621475836
1,25368553,rs7527379,1,28208168,rs12140070,0.1645813527877886</code></pre><pre><code class="language-julia"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;ordinalgwas.null.txt&quot;, force=true)</code></pre><h2 id="Plotting-Results"><a class="docs-heading-anchor" href="#Plotting-Results">Plotting Results</a><a id="Plotting-Results-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting-Results" title="Permalink"></a></h2><p>To plot the GWAS results, use the <a href="https://openmendel.github.io/MendelPlots.jl/latest/">MendelPlots.jl package</a>.</p><h2 id="Docker"><a class="docs-heading-anchor" href="#Docker">Docker</a><a id="Docker-1"></a><a class="docs-heading-anchor-permalink" href="#Docker" title="Permalink"></a></h2><p>For ease of using OrdinalGWAS, we provide a Dockerfile so users don&#39;t need to install Julia and required packages. Only Docker app needs to be installed in order to run analysis. Following is tested on Docker 2.0.0.0-mac78.</p><p><strong>Step 1</strong>: Create a Dockerfile with content <a href="https://raw.githubusercontent.com/OpenMendel/OrdinalGWAS.jl/master/docker/Dockerfile">here</a>, or, if the bash command <code>wget</code> is available, obtain Dockerfile by</p><pre><code class="language-bash"># on command line
wget https://raw.githubusercontent.com/OpenMendel/OrdinalGWAS.jl/master/docker/Dockerfile</code></pre><p><strong>Step 2</strong>: Build a docker image called <code>ordinalgwas-app</code>, assuming that the Dockerfile is located in the <code>../docker</code> folder. Building the image for the first time can take up to 10 minutes; but it only needs to be done once.</p><pre><code class="language-bash"># on command line
docker build -t ordinalgwas-app ../docker/</code></pre><p><strong>Step 3</strong>: Suppose data files are located at <code>/path/to/data</code> folder, run analysis by</p><pre><code class="language-bash"># on command line
docker run -v /path/to/data:/data -t ordinalgwas-app julia -e &#39;using OrdinalGWAS; ordinalgwas(@formula(trait ~ sex), &quot;/data/covariate.txt&quot;, &quot;/data/hapmap3&quot;, nullfile=&quot;/data/ordinalgwas.null.txt&quot;, pvalfile=&quot;/data/ordinalgwas&quot;);&#39;</code></pre><p>Here  </p><ul><li><code>-t ordinalgwas-app</code> creates a container using the <code>ordinalgwas-app</code> image build in step 2.  </li><li><code>-v /path/to/data:/data</code> maps the <code>/path/to/data</code> folder on host machine to the <code>/data</code> folder within the container. </li><li><code>julia -e &#39;using OrdinalGWAS; ordinalgwas(@formula(trait ~ sex), &quot;/data/covariate.txt&quot;, &quot;/data/hapmap3&quot;, nullfile=&quot;/data/ordinalgwas.null.txt&quot;, pvalfile=&quot;/data/ordinalgwas&quot;);</code> calls Julia and runs <code>ordinalgwas</code> function. </li></ul><p>The output files are written in <code>/path/to/data</code> directory.</p><h2 id="Multiple-Plink-file-sets"><a class="docs-heading-anchor" href="#Multiple-Plink-file-sets">Multiple Plink file sets</a><a id="Multiple-Plink-file-sets-1"></a><a class="docs-heading-anchor-permalink" href="#Multiple-Plink-file-sets" title="Permalink"></a></h2><p>In large scale studies, genotypes data are split into multiple Plink files, e.g., by chromosome. Then GWAS analysis can be done in parallel. This can be achieved by two steps.</p><p>Let&#39;s first create demo data by splitting hapmap3 according to chromosome:</p><pre><code class="language-julia"># split example hapmap3 data according to chromosome
SnpArrays.split_plink(datadir * &quot;hapmap3&quot;, :chromosome; prefix=datadir * &quot;hapmap3.chr.&quot;)
readdir(glob&quot;hapmap3.chr.*&quot;, datadir)</code></pre><pre><code class="language-none">┌ Warning: `getindex(df::DataFrame, col_ind::ColumnIndex)` is deprecated, use `df[!, col_ind]` instead.
│   caller = #split_plink#77(::String, ::typeof(split_plink), ::SnpData, ::Symbol) at snpdata.jl:94
└ @ SnpArrays /Users/christophergerman/.julia/packages/SnpArrays/d0iJw/src/snpdata.jl:94





75-element Array{String,1}:
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.1.bed&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.1.bim&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.1.fam&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.10.bed&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.10.bim&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.10.fam&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.11.bed&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.11.bim&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.11.fam&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.12.bed&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.12.bim&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.12.fam&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.13.bed&quot;
 ⋮                                                                        
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.6.bed&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.6.bim&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.6.fam&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.7.bed&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.7.bim&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.7.fam&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.8.bed&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.8.bim&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.8.fam&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.9.bed&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.9.bim&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.9.fam&quot;</code></pre><p>Step 1: Fit the null model. Setting third argument <code>plinkfile</code> to <code>nothing</code> instructs <code>ordinalgwas</code> function to fit the null model only.</p><pre><code class="language-julia">nm = ordinalgwas(@formula(trait ~ sex), datadir * &quot;covariate.txt&quot;, nothing)</code></pre><pre><code class="language-none">StatsModels.TableRegressionModel{OrdinalMultinomialModel{Int64,Float64,LogitLink},Array{Float64,2}}

trait ~ sex

Coefficients:
──────────────────────────────────────────────────────
               Estimate  Std.Error   t value  Pr(&gt;|t|)
──────────────────────────────────────────────────────
intercept1|2  -1.48564    0.358891  -4.13952    &lt;1e-4 
intercept2|3  -0.569479   0.341044  -1.66981    0.0959
intercept3|4   0.429815   0.339642   1.26549    0.2066
sex            0.424656   0.213914   1.98517    0.0480
──────────────────────────────────────────────────────</code></pre><p>Step 2: GWAS for each chromosome.</p><pre><code class="language-julia"># this part can be submitted as separate jobs
for chr in 1:23
    plinkfile = datadir * &quot;hapmap3.chr.&quot; * string(chr)
    pvalfile = plinkfile * &quot;.pval.txt&quot; 
    ordinalgwas(nm, plinkfile, pvalfile=pvalfile)
end</code></pre><pre><code class="language-julia"># show the result files
readdir(glob&quot;*.pval.txt&quot;, datadir)</code></pre><pre><code class="language-none">23-element Array{String,1}:
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.1.pval.txt&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.10.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.11.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.12.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.13.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.14.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.15.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.16.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.17.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.18.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.19.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.2.pval.txt&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.20.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.21.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.22.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.23.pval.txt&quot;
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.3.pval.txt&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.4.pval.txt&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.5.pval.txt&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.6.pval.txt&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.7.pval.txt&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.8.pval.txt&quot; 
 &quot;/Users/christophergerman/.julia/dev/OrdinalGWAS/data/hapmap3.chr.9.pval.txt&quot;</code></pre><p>In the rare situations where the multiple sets of Plink files lack the <code>fam</code> file or the corresponding bed and bim files have different filenames, users can explicitly supply bed filename, bim file name, and number of individuals. Replace Step 2 by </p><p>Step 2&#39;: GWAS for each chromosome.</p><pre><code class="language-julia"># this part can be submitted as separate jobs
for chr in 1:23
    bedfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.bed&quot;
    bimfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.bim&quot;
    pvalfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.pval.txt&quot;
    ordinalgwas(nm, bedfile, bimfile, 324; pvalfile=pvalfile)
end</code></pre><pre><code class="language-julia"># clean up
rm(&quot;ordinalgwas.null.txt&quot;, force=true)
isfile(datadir * &quot;fittednullmodel.jld2&quot;) &amp;&amp; rm(datadir * &quot;fittednullmodel.jld2&quot;)
for chr in 1:23
    pvalfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.pval.txt&quot;
    rm(pvalfile, force=true)
end
for chr in 1:26
    plinkfile = datadir * &quot;hapmap3.chr.&quot; * string(chr)
    rm(plinkfile * &quot;.bed&quot;, force=true)
    rm(plinkfile * &quot;.fam&quot;, force=true)
    rm(plinkfile * &quot;.bim&quot;, force=true)
end</code></pre><h2 id="Multiple-Plink-file-sets-on-cluster"><a class="docs-heading-anchor" href="#Multiple-Plink-file-sets-on-cluster">Multiple Plink file sets on cluster</a><a id="Multiple-Plink-file-sets-on-cluster-1"></a><a class="docs-heading-anchor-permalink" href="#Multiple-Plink-file-sets-on-cluster" title="Permalink"></a></h2><p>We provide two scripts that successfully run on UCLA&#39;s Hoffman2 cluster using Julia v1.0.1 and PBS job schedulaer (<code>qsub</code>).</p><ul><li>The first script <a href="https://raw.githubusercontent.com/OpenMendel/OrdinalGWAS.jl/master/docs/cluster_preparedata.jl"><code>cluster_preparedata.jl</code></a> creates a demo data set in current folder. Run</li></ul><pre><code class="language-julia">julia cluster_preparedata.jl</code></pre><p>on head node.</p><pre><code class="language-julia">run(`cat cluster_preparedata.jl`);</code></pre><pre><code class="language-none">#!/usr/local/bin/julia
#
# This script prepares a data set in current folder. 
# For each of chromosome 1-23, there is a set gzipped Plink files:
# hapmap3.chr.1.bed.gz, hapmap3.chr.1.bim.gz, hapmap3.chr.1.fam.gz
# hapmap3.chr.2.bed.gz, hapmap3.chr.2.bim.gz, hapmap3.chr.2.fam.gz
# ...
# hapmap3.chr.23.bed.gz, hapmap3.chr.23.bim.gz, hapmap3.chr.23.fam.gz
# There is also a csv file &quot;covariate.txt&quot; that contains trait and covariates.
#

# install and load Julia packages
using Pkg
if haskey(Pkg.installed(), &quot;SnpArrays&quot;)
    Pkg.update(&quot;SnpArrays&quot;)
else
    Pkg.add(PackageSpec(url=&quot;https://github.com/OpenMendel/SnpArrays.jl.git&quot;))
end
if haskey(Pkg.installed(), &quot;OrdinalMultinomialModels&quot;)
    Pkg.update(&quot;OrdinalMultinomialModels&quot;)
else
    Pkg.add(PackageSpec(url=&quot;https://github.com/OpenMendel/OrdinalMultinomialModels.jl.git&quot;))
end
if haskey(Pkg.installed(), &quot;OrdinalGWAS&quot;)
    Pkg.update(&quot;OrdinalGWAS&quot;)
else
    Pkg.add(PackageSpec(url=&quot;https://github.com/OpenMendel/OrdinalGWAS.jl.git&quot;))
end
using OrdinalMultinomialModels, OrdinalGWAS, SnpArrays

# split hapmap3 data according to chromosome
datadir = normpath(joinpath(dirname(pathof(OrdinalGWAS)), &quot;../data/&quot;))
SnpArrays.split_plink(datadir * &quot;hapmap3&quot;, :chromosome; prefix = &quot;hapmap3.chr.&quot;)
# compresse Plink files for chromosome 1-23
for chr in 1:23
    plinkfile = &quot;hapmap3.chr.&quot; * string(chr)
    SnpArrays.compress_plink(plinkfile)
end
# delete uncompressed chromosome Plink files
for chr in 1:26
    plinkfile = &quot;hapmap3.chr.&quot; * string(chr)
    rm(plinkfile * &quot;.bed&quot;, force=true)
    rm(plinkfile * &quot;.bim&quot;, force=true)
    rm(plinkfile * &quot;.fam&quot;, force=true)
end
# copy covariate.txt file
cp(datadir * &quot;covariate.txt&quot;, joinpath(pwd(), &quot;covariate.txt&quot;))</code></pre><ul><li>The second script <a href="https://raw.githubusercontent.com/OpenMendel/OrdinalGWAS.jl/master/docs/cluster_run.jl"><code>cluster_run.jl</code></a> first fits the null model then submits a separate job for each chromosome. Run</li></ul><pre><code class="language-julia">julia cluster_run.jl</code></pre><p>on head node.</p><pre><code class="language-julia">run(`cat cluster_run.jl`);</code></pre><pre><code class="language-none">#!/usr/local/bin/julia
#
# This script demonstrates how to submit multiple OrdinalGWAS runs from multiple sets of
# Plink files on UCLA Hoffman2 cluster. It assumes that a demo data is available by
# running `julia cluster_preparedata.jl` at current folder.
#

using OrdinalGWAS, Serialization

# Step 1: fit null model and save result to file `fittednullmodel.jls`
nm = ordinalgwas(@formula(trait ~ sex), &quot;covariate.txt&quot;, nothing)
open(&quot;fittednullmodel.jls&quot;, &quot;w&quot;) do io
    Serialization.serialize(io, nm)
end

# Step 2: GWAS for each chromosome
for chr in 1:23
    println(&quot;submit job for chromosome=$chr&quot;)
    jcode = &quot;using OrdinalGWAS, Serialization;
    nm = open(deserialize, \&quot;fittednullmodel.jls\&quot;);
    bedfile = \&quot;hapmap3.chr.\&quot; * string($chr) * \&quot;.bed.gz\&quot;;
    bimfile = \&quot;hapmap3.chr.\&quot; * string($chr) * \&quot;.bim.gz\&quot;;
    pvalfile = \&quot;hapmap3.chr.\&quot; * string($chr) * \&quot;.pval.txt\&quot;;
    ordinalgwas(nm, bedfile, bimfile, 324; pvalfile=pvalfile);&quot;
    # prepare sh file for qsub
    open(&quot;tmp.sh&quot;, &quot;w&quot;) do io
        println(io, &quot;#!/bin/bash&quot;)
        println(io, &quot;#\$ -cwd&quot;)
        println(io, &quot;# error = Merged with joblog&quot;)
        println(io, &quot;#\$ -o joblog.\$JOB_ID&quot;)
        println(io, &quot;#\$ -j y&quot;)
        println(io, &quot;#\$ -l h_rt=0:30:00,h_data=2G&quot;) # request runtime and memory
        println(io, &quot;#\$ -pe shared 2&quot;) # request # shared-memory nodes
        println(io, &quot;# Email address to notify&quot;)
        println(io, &quot;#\$ -M \$USER@mail&quot;)
        println(io, &quot;# Notify when&quot;)
        println(io, &quot;#\$ -m a&quot;)
        println(io)
        println(io, &quot;# load the job environment:&quot;)
        println(io, &quot;. /u/local/Modules/default/init/modules.sh&quot;)
        println(io, &quot;module load julia/1.0.1&quot;) # available Julia version
        println(io)
        println(io, &quot;# run julia code&quot;)
        println(io, &quot;julia -e &#39;$jcode&#39; &gt; output.\$JOB_ID 2&gt;&amp;1&quot;)
    end
    # submit job
    run(`qsub tmp.sh`)
end</code></pre><pre><code class="language-julia"></code></pre></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 21 July 2020 01:32">Tuesday 21 July 2020</span>. Using Julia version 1.4.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
